{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load all article metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/scott/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (4,5,10) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "# Load database of available articles\n",
    "df = pd.read_csv('/gh/data/opencode/PMC-ids.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Journal Title</th>\n",
       "      <th>ISSN</th>\n",
       "      <th>eISSN</th>\n",
       "      <th>Year</th>\n",
       "      <th>Volume</th>\n",
       "      <th>Issue</th>\n",
       "      <th>Page</th>\n",
       "      <th>DOI</th>\n",
       "      <th>PMCID</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Manuscript Id</th>\n",
       "      <th>Release Date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Breast Cancer Res</td>\n",
       "      <td>1465-5411</td>\n",
       "      <td>1465-542X</td>\n",
       "      <td>2000</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMC13900</td>\n",
       "      <td>11250746.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>live</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Breast Cancer Res</td>\n",
       "      <td>1465-5411</td>\n",
       "      <td>1465-542X</td>\n",
       "      <td>2000</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMC13901</td>\n",
       "      <td>11250747.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>live</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Breast Cancer Res</td>\n",
       "      <td>1465-5411</td>\n",
       "      <td>1465-542X</td>\n",
       "      <td>2000</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>66</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMC13902</td>\n",
       "      <td>11250748.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>live</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Breast Cancer Res</td>\n",
       "      <td>1465-5411</td>\n",
       "      <td>1465-542X</td>\n",
       "      <td>1999</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>10.1186/bcr29</td>\n",
       "      <td>PMC13911</td>\n",
       "      <td>11056684.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>live</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Breast Cancer Res</td>\n",
       "      <td>1465-5411</td>\n",
       "      <td>1465-542X</td>\n",
       "      <td>1999</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>64</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMC13912</td>\n",
       "      <td>11400682.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>live</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Journal Title       ISSN      eISSN  Year Volume Issue Page  \\\n",
       "0  Breast Cancer Res  1465-5411  1465-542X  2000      3     1   55   \n",
       "1  Breast Cancer Res  1465-5411  1465-542X  2000      3     1   61   \n",
       "2  Breast Cancer Res  1465-5411  1465-542X  2000      3     1   66   \n",
       "3  Breast Cancer Res  1465-5411  1465-542X  1999      2     1   59   \n",
       "4  Breast Cancer Res  1465-5411  1465-542X  1999      2     1   64   \n",
       "\n",
       "             DOI     PMCID        PMID Manuscript Id Release Date  \n",
       "0            NaN  PMC13900  11250746.0           NaN         live  \n",
       "1            NaN  PMC13901  11250747.0           NaN         live  \n",
       "2            NaN  PMC13902  11250748.0           NaN         live  \n",
       "3  10.1186/bcr29  PMC13911  11056684.0           NaN         live  \n",
       "4            NaN  PMC13912  11400682.0           NaN         live  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How many journals at least 500 articles since 2014?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "631\n"
     ]
    }
   ],
   "source": [
    "df_2014 = df[df['Year'] >= 2014]\n",
    "js_count = df_2014['Journal Title'].value_counts()\n",
    "js_use = js_count[js_count > 500]\n",
    "print(len(js_use))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "journals_big = ['PLoS One', 'Sci Rep', 'Proc Natl Acad Sci U S A', 'Nat Commun']\n",
    "journals_smaller = ['eLife', 'J Neurosci', 'Front Hum Neurosci', 'Front Neurosci',\n",
    "                'F1000Res', 'J Neurophysiol', 'Nature', 'Neuroimage', 'Neuron', 'Science',\n",
    "                'Brain', 'eNeuro', 'Neuroscience', 'Front Syst Neurosci', 'Sci Data'\n",
    "                'Nat Neurosci', 'Nat Methods']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Demo scrape some full text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "apikey = open('apikey.txt', 'r').read()\n",
    "db = 'pmc'\n",
    "paper_id = 'PMC3877685'\n",
    "base = 'http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?'\n",
    "s = '{:s}db={:s}&id={:s}&api_key={:s}'.format(base, db, paper_id, apikey)\n",
    "out = requests.get(s)\n",
    "page_soup = BeautifulSoup(out.content, 'lxml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<?xml version=\"1.0\" ?><!DOCTYPE pmc-articleset PUBLIC \"-//NLM//DTD ARTICLE SET 2.0//EN\" \"https://dtd.nlm.nih.gov/ncbi/pmc/articleset/nlm-articleset-2.0.dtd\">\n",
       "<html><body><pmc-articleset><article article-type=\"research-article\" xmlns:mml=\"http://www.w3.org/1998/Math/MathML\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<?properties open_access?>\n",
       "<front>\n",
       "<journal-meta>\n",
       "<journal-id journal-id-type=\"nlm-ta\">PLoS One</journal-id>\n",
       "<journal-id journal-id-type=\"iso-abbrev\">PLoS ONE</journal-id>\n",
       "<journal-id journal-id-type=\"publisher-id\">plos</journal-id>\n",
       "<journal-id journal-id-type=\"pmc\">plosone</journal-id>\n",
       "<journal-title-group>\n",
       "<journal-title>PLoS ONE</journal-title>\n",
       "</journal-title-group>\n",
       "<issn pub-type=\"epub\">1932-6203</issn>\n",
       "<publisher>\n",
       "<publisher-name>Public Library of Science</publisher-name>\n",
       "<publisher-loc>San Francisco, USA</publisher-loc>\n",
       "</publisher>\n",
       "</journal-meta>\n",
       "<article-meta>\n",
       "<article-id pub-id-type=\"pmid\">24392115</article-id>\n",
       "<article-id pub-id-type=\"pmc\">3877685</article-id>\n",
       "<article-id pub-id-type=\"publisher-id\">PONE-D-13-37281</article-id>\n",
       "<article-id pub-id-type=\"doi\">10.1371/journal.pone.0084227</article-id>\n",
       "<article-categories>\n",
       "<subj-group subj-group-type=\"heading\">\n",
       "<subject>Research Article</subject>\n",
       "</subj-group>\n",
       "<subj-group subj-group-type=\"Discipline-v2\">\n",
       "<subject>Biology</subject>\n",
       "<subj-group>\n",
       "<subject>Computational Biology</subject>\n",
       "<subj-group>\n",
       "<subject>Molecular Genetics</subject>\n",
       "<subj-group>\n",
       "<subject>Gene Classes</subject>\n",
       "<subject>Gene Identification and Analysis</subject>\n",
       "<subject>Gene Expression</subject>\n",
       "</subj-group>\n",
       "</subj-group>\n",
       "<subj-group>\n",
       "<subject>Metabolic Networks</subject>\n",
       "<subject>Microarrays</subject>\n",
       "<subject>Regulatory Networks</subject>\n",
       "<subject>Signaling Networks</subject>\n",
       "</subj-group>\n",
       "</subj-group>\n",
       "<subj-group>\n",
       "<subject>Genetics</subject>\n",
       "<subj-group>\n",
       "<subject>Human Genetics</subject>\n",
       "<subj-group>\n",
       "<subject>Genetic Association Studies</subject>\n",
       "</subj-group>\n",
       "</subj-group>\n",
       "<subj-group>\n",
       "<subject>Gene Networks</subject>\n",
       "<subject>Genetics of Disease</subject>\n",
       "</subj-group>\n",
       "</subj-group>\n",
       "</subj-group>\n",
       "<subj-group subj-group-type=\"Discipline-v2\">\n",
       "<subject>Computer Science</subject>\n",
       "<subj-group>\n",
       "<subject>Algorithms</subject>\n",
       "</subj-group>\n",
       "</subj-group>\n",
       "</article-categories>\n",
       "<title-group>\n",
       "<article-title>Automatic Context-Specific Subnetwork Discovery from Large Interaction Networks</article-title>\n",
       "<alt-title alt-title-type=\"running-head\">Context-Specific Subnetwork Discovery</alt-title>\n",
       "</title-group>\n",
       "<contrib-group>\n",
       "<contrib contrib-type=\"author\">\n",
       "<name>\n",
       "<surname>Saha</surname>\n",
       "<given-names>Ashis</given-names>\n",
       "</name>\n",
       "<xref ref-type=\"aff\" rid=\"aff1\">\n",
       "<sup>1</sup>\n",
       "</xref>\n",
       "</contrib>\n",
       "<contrib contrib-type=\"author\">\n",
       "<name>\n",
       "<surname>Tan</surname>\n",
       "<given-names>Aik Choon</given-names>\n",
       "</name>\n",
       "<xref ref-type=\"aff\" rid=\"aff2\">\n",
       "<sup>2</sup>\n",
       "</xref>\n",
       "</contrib>\n",
       "<contrib contrib-type=\"author\">\n",
       "<name>\n",
       "<surname>Kang</surname>\n",
       "<given-names>Jaewoo</given-names>\n",
       "</name>\n",
       "<xref ref-type=\"aff\" rid=\"aff1\">\n",
       "<sup>1</sup>\n",
       "</xref>\n",
       "<xref ref-type=\"aff\" rid=\"aff3\">\n",
       "<sup>3</sup>\n",
       "</xref>\n",
       "<xref ref-type=\"corresp\" rid=\"cor1\">\n",
       "<sup>*</sup>\n",
       "</xref>\n",
       "</contrib>\n",
       "</contrib-group>\n",
       "<aff id=\"aff1\">\n",
       "<label>1</label>\n",
       "<addr-line>Department of Computer Science and Engineering, Korea University, Seoul, Korea</addr-line>\n",
       "</aff>\n",
       "<aff id=\"aff2\">\n",
       "<label>2</label>\n",
       "<addr-line>Department of Medicine/Medical Oncology, University of Colorado Anschutz Medical Campus, Aurora, Colorado, United States of America</addr-line>\n",
       "</aff>\n",
       "<aff id=\"aff3\">\n",
       "<label>3</label>\n",
       "<addr-line>Interdisciplinary Graduate Program in Bioinformatics, Korea University, Seoul, Korea</addr-line>\n",
       "</aff>\n",
       "<contrib-group>\n",
       "<contrib contrib-type=\"editor\">\n",
       "<name>\n",
       "<surname>Aloy</surname>\n",
       "<given-names>Patrick</given-names>\n",
       "</name>\n",
       "<role>Editor</role>\n",
       "<xref ref-type=\"aff\" rid=\"edit1\"></xref>\n",
       "</contrib>\n",
       "</contrib-group>\n",
       "<aff id=\"edit1\">\n",
       "<addr-line>Institute for Research in Biomedicine, Spain</addr-line>\n",
       "</aff>\n",
       "<author-notes>\n",
       "<corresp id=\"cor1\">* E-mail: <email>kangj@korea.ac.kr</email></corresp>\n",
       "<fn fn-type=\"COI-statement\">\n",
       "<p><bold>Competing Interests: </bold>The authors have declared that no competing interests exist.</p>\n",
       "</fn>\n",
       "<fn fn-type=\"con\">\n",
       "<p>Conceived and designed the experiments: AS JK. Performed the experiments: AS. Analyzed the data: AS ACT JK. Wrote the paper: AS ACT JK.</p>\n",
       "</fn>\n",
       "</author-notes>\n",
       "<pub-date pub-type=\"collection\">\n",
       "<year>2014</year>\n",
       "</pub-date>\n",
       "<pub-date pub-type=\"epub\">\n",
       "<day>1</day>\n",
       "<month>1</month>\n",
       "<year>2014</year>\n",
       "</pub-date>\n",
       "<volume>9</volume>\n",
       "<issue>1</issue>\n",
       "<elocation-id>e84227</elocation-id>\n",
       "<history>\n",
       "<date date-type=\"received\">\n",
       "<day>10</day>\n",
       "<month>9</month>\n",
       "<year>2013</year>\n",
       "</date>\n",
       "<date date-type=\"accepted\">\n",
       "<day>21</day>\n",
       "<month>11</month>\n",
       "<year>2013</year>\n",
       "</date>\n",
       "</history>\n",
       "<permissions>\n",
       "<copyright-statement>© 2014 Saha et al</copyright-statement>\n",
       "<copyright-year>2014</copyright-year>\n",
       "<copyright-holder>Saha et al</copyright-holder>\n",
       "<license xlink:href=\"http://creativecommons.org/licenses/by/4.0/\">\n",
       "<license-p>This is an open-access article distributed under the terms of the Creative Commons Attribution License, which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are properly credited.</license-p>\n",
       "</license>\n",
       "</permissions>\n",
       "<abstract>\n",
       "<p>Genes act in concert via specific networks to drive various biological processes, including progression of diseases such as cancer. Under different phenotypes, different subsets of the gene members of a network participate in a biological process. Single gene analyses are less effective in identifying such core gene members (subnetworks) within a gene set/network, as compared to gene set/network-based analyses. Hence, it is useful to identify a discriminative classifier by focusing on the subnetworks that correspond to different phenotypes. Here we present a novel algorithm to automatically discover the important subnetworks of closely interacting molecules to differentiate between two phenotypes (context) using gene expression profiles. We name it COSSY (COntext-Specific Subnetwork discoverY). It is a non-greedy algorithm and thus unlikely to have local optima problems. COSSY works for any interaction network regardless of the network topology. One added benefit of COSSY is that it can also be used as a highly accurate classification platform which can produce a set of interpretable features.</p>\n",
       "</abstract>\n",
       "<funding-group>\n",
       "<funding-statement>This work was supported by the National Research Foundation of Korea (NRF) grant funded by the Korean government (MSIP) (2012R1A2A2A01014729); and Next-Generation Information Computing Development Program through the National Research Foundation of Korea (NRF) funded by the Ministry of Science, ICT &amp; Future Planning (2012M3C4A7033341). The funders had no role in study design, data collection and analysis, decision to publish, or preparation of the manuscript.</funding-statement>\n",
       "</funding-group>\n",
       "<counts>\n",
       "<page-count count=\"12\"></page-count>\n",
       "</counts>\n",
       "</article-meta>\n",
       "</front>\n",
       "<sec id=\"s1\">\n",
       "<title>Introduction</title>\n",
       "<p>Biological systems are complex in nature; individual molecular components (e.g., genes and proteins) interact with each other in specific networks to exert their functions. The development of new technologies and powerful computational algorithms to sequence and characterize genomes have enabled researchers to acquire and analyze tens of thousands of ‘omic’ data points across the genetic and epigenetic changes within biological systems. One of the main difficulties in the gene expression analysis is the size of the ‘omic’ data. Due to the large number of genes, the complexity of the algorithms becomes very high.</p>\n",
       "<p>The common practice is to rank the individual genes according to a similarity or dissimilarity score and then use the important genes for prediction. However, Risch (2000) suggested that a combination of genes is more apt for common disease classification than individual genes <xref ref-type=\"bibr\" rid=\"pone.0084227-Risch1\">[1]</xref>. Marchini, Donnelly, and Cardon (2005) illustrated the effect that gene-gene interactions have on complex diseases <xref ref-type=\"bibr\" rid=\"pone.0084227-Marchini1\">[2]</xref>. Moreover, recent research suggests that molecular networks, rather than random sets of genes, can help find underlying physiological states associated with disease and understand complex biological processes <xref ref-type=\"bibr\" rid=\"pone.0084227-Schadt1\">[3]</xref>. Such approaches are more objective and robust in their ability to discover sets of coordinated differentially expressed genes among pathway members and their association to a specific biological phenotype. These analyses may provide new insights linking biological phenotypes to their underlying molecular mechanisms, as well as suggesting new hypotheses about the pathway membership and connectivity.</p>\n",
       "<p>For this reason, researchers are now concentrating on gene-gene interactions instead of on individual gene analyses. Geman et al. (2004) introduced the relative expression reversal concept of top-scoring pair (TSP) as a classification approach for gene expression profiles <xref ref-type=\"bibr\" rid=\"pone.0084227-Geman1\">[4]</xref>. They attempted to figure out the pairs of genes whose expression levels typically invert from one class to another. This concept has been improved by majority voting of the <inline-formula><inline-graphic xlink:href=\"pone.0084227.e001.jpg\"></inline-graphic></inline-formula> top scoring pairs (<italic>k</italic>-TSP) <xref ref-type=\"bibr\" rid=\"pone.0084227-Tan1\">[5]</xref>. Chopra et al. (2010) also introduced their version of gene-pairing which they call <italic>doublet</italic>\n",
       "<xref ref-type=\"bibr\" rid=\"pone.0084227-Chopra1\">[6]</xref>. Instead of using only the inverse relations of the expression levels, they provided a framework for using different kinds of relations between two genes. Methods that extended the same concept for finding triplets <xref ref-type=\"bibr\" rid=\"pone.0084227-Lin1\">[7]</xref> and other combinations of gene pairs <xref ref-type=\"bibr\" rid=\"pone.0084227-Magis1\">[8]</xref>, <xref ref-type=\"bibr\" rid=\"pone.0084227-Wang1\">[9]</xref> have been developed.</p>\n",
       "<p>Although these methods produce highly accurate results, they only consider the interaction of a few genes at a time which is still a small number to imply the biological process of a disease. Conversely, gene set enrichment analysis methods such as the Gene Set Enrichment Analysis (GSEA) <xref ref-type=\"bibr\" rid=\"pone.0084227-Subramanian1\">[10]</xref> have improved the ability to identify candidate genes that are correlated with a disease state by exploiting the idea that gene expression alterations might be revealed at the level of biological pathways or co-regulated gene sets, rather than at the level of individual genes. However, the interactions (or topology) of the identified subset of genes within the predefined gene sets are not captured by these methods.</p>\n",
       "<p>The availability of large-scale molecular interaction data such as those collected in KEGG (Kyoto Encyclopedia of Genes and Genomes) <xref ref-type=\"bibr\" rid=\"pone.0084227-Kanehisa1\">[11]</xref> and STRING (Search Tool for the Retrieval of Interacting Genes/Proteins) <xref ref-type=\"bibr\" rid=\"pone.0084227-Franceschini1\">[12]</xref> provides an opportunity to incorporate these interactions into omic data analysis. The advantage of these network-based approaches is that they can fully describe the complex biological processes driving disease phenotypes. Most network-based approaches attempted to identify informative subnetworks - a set of genes that interact directly and have differential expression patterns - for disease classification. These approaches attempted to find these subnetworks using a greedy approach, starting with a small network (typically a single node) and adding nodes greedily until the score is improved <xref ref-type=\"bibr\" rid=\"pone.0084227-Chuang1\">[13]</xref>–<xref ref-type=\"bibr\" rid=\"pone.0084227-Su1\">[16]</xref>. However, they may suffer from local-optima problems and may not even work for a sparse network such as KEGG because their greedy expansion process could halt imgmaturely due to sparse connectivity. Alternatively, some approaches, such as DIRAC (Differential Rank Conservation)<xref ref-type=\"bibr\" rid=\"pone.0084227-Eddy1\">[17]</xref>, utilized user-defined and pre-partitioned subnetworks for classification.</p>\n",
       "<p>Here we propose a novel COntext-Specific Subnetwork discoverY (COSSY) algorithm to automatically discover the important subnetworks to differentiate between two phenotypes (context). COSSY finds the important subnetworks using a non-greedy approach. Unlike the greedy approach, our approach first divides the network, and then ranks the subnetworks based on the biological context of interest. We demonstrate that COSSY works for any network regardless of its topology. We show that COSSY is capable of identifying discriminative subnetworks while achieving classification results comparable to those of other state-of-the-art methods. Finally, we illustrate that the subnetworks identified by COSSY provide meaningful biological insights from high-throughput datasets.</p>\n",
       "<sec id=\"s1a\">\n",
       "<title>COSSY Algorithm</title>\n",
       "<p>Let <inline-formula><inline-graphic xlink:href=\"pone.0084227.e002.jpg\"></inline-graphic></inline-formula> be a molecular interaction network (graph) where each vertex in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e003.jpg\"></inline-graphic></inline-formula> denotes a molecule (or a set of molecules) and each edge in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e004.jpg\"></inline-graphic></inline-formula> denotes an interaction between molecules. A molecule can be a gene, a protein, a nucleic acid or a similar biochemical object. Let <inline-formula><inline-graphic xlink:href=\"pone.0084227.e005.jpg\"></inline-graphic></inline-formula> be the training dataset containing the microarray expression profiles of <inline-formula><inline-graphic xlink:href=\"pone.0084227.e006.jpg\"></inline-graphic></inline-formula> samples, where each profile consists of <inline-formula><inline-graphic xlink:href=\"pone.0084227.e007.jpg\"></inline-graphic></inline-formula> probes. <inline-formula><inline-graphic xlink:href=\"pone.0084227.e008.jpg\"></inline-graphic></inline-formula> can be represented as an <inline-formula><inline-graphic xlink:href=\"pone.0084227.e009.jpg\"></inline-graphic></inline-formula> matrix where <inline-formula><inline-graphic xlink:href=\"pone.0084227.e010.jpg\"></inline-graphic></inline-formula> denotes the expression value of the <inline-formula><inline-graphic xlink:href=\"pone.0084227.e011.jpg\"></inline-graphic></inline-formula>-th probe, <inline-formula><inline-graphic xlink:href=\"pone.0084227.e012.jpg\"></inline-graphic></inline-formula>, in the <inline-formula><inline-graphic xlink:href=\"pone.0084227.e013.jpg\"></inline-graphic></inline-formula>-th sample, <inline-formula><inline-graphic xlink:href=\"pone.0084227.e014.jpg\"></inline-graphic></inline-formula>. The row vector <inline-formula><inline-graphic xlink:href=\"pone.0084227.e015.jpg\"></inline-graphic></inline-formula> thus represents the <inline-formula><inline-graphic xlink:href=\"pone.0084227.e016.jpg\"></inline-graphic></inline-formula>-th probe across <inline-formula><inline-graphic xlink:href=\"pone.0084227.e017.jpg\"></inline-graphic></inline-formula> samples and the column vector <inline-formula><inline-graphic xlink:href=\"pone.0084227.e018.jpg\"></inline-graphic></inline-formula> represents the <inline-formula><inline-graphic xlink:href=\"pone.0084227.e019.jpg\"></inline-graphic></inline-formula>-th sample across <inline-formula><inline-graphic xlink:href=\"pone.0084227.e020.jpg\"></inline-graphic></inline-formula> probes. Let <inline-formula><inline-graphic xlink:href=\"pone.0084227.e021.jpg\"></inline-graphic></inline-formula> be the vector of the class labels (context) for <inline-formula><inline-graphic xlink:href=\"pone.0084227.e022.jpg\"></inline-graphic></inline-formula> samples where <inline-formula><inline-graphic xlink:href=\"pone.0084227.e023.jpg\"></inline-graphic></inline-formula>. Our task is to find the subnetworks (subgraphs) of <inline-formula><inline-graphic xlink:href=\"pone.0084227.e024.jpg\"></inline-graphic></inline-formula> to predict the class (context) of a new sample <inline-formula><inline-graphic xlink:href=\"pone.0084227.e025.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>To give an overview, we first partition the molecular interaction network <inline-formula><inline-graphic xlink:href=\"pone.0084227.e026.jpg\"></inline-graphic></inline-formula> to generate the molecular interaction subnetworks. Each subnetwork is then ranked using a clustering approach and finally the top <inline-formula><inline-graphic xlink:href=\"pone.0084227.e027.jpg\"></inline-graphic></inline-formula> subnetworks vote to predict the context of the new sample, <inline-formula><inline-graphic xlink:href=\"pone.0084227.e028.jpg\"></inline-graphic></inline-formula> (<xref ref-type=\"fig\" rid=\"pone-0084227-g001\">Figure 1</xref>).</p>\n",
       "<fig id=\"pone-0084227-g001\" orientation=\"portrait\" position=\"float\">\n",
       "<object-id pub-id-type=\"doi\">10.1371/journal.pone.0084227.g001</object-id>\n",
       "<label>Figure 1</label>\n",
       "<caption>\n",
       "<title>Overview of COSSY.</title>\n",
       "<p>Communities are extracted from the molecular interaction network to generate Molecular Interaction Subnetworks (MISs)(A–B). Each MIS is mapped to microarray probes and all the samples are clustered according to the expression pattern of a certain number of highly differentially expressed probes (3 probes in this example figure). MISs are ranked by the entropy score which is lowest when every cluster contains only one type (phenotype) of samples (C–D). Finally, the top <inline-formula><inline-graphic xlink:href=\"pone.0084227.e029.jpg\"></inline-graphic></inline-formula> MISs cast votes to predict the context (phenotype) of a new sample. The voting depends on the proportion of different types of samples in the cluster closest to the new sample (E–G).</p>\n",
       "</caption>\n",
       "<graphic xlink:href=\"pone.0084227.g001\"></graphic>\n",
       "</fig>\n",
       "</sec>\n",
       "<sec id=\"s1b\">\n",
       "<title>Molecular Interaction Subnetwork (MIS)</title>\n",
       "<p>We use a biological knowledge-base in the form of a network in order to get biologically interpretable results. An important property of such a network is that it shows how a group of molecules interact with each other, often performing a molecular function. However, it is not necessary for all the molecules in a network to interact with each other. For example, some of the genes in the thyroid cancer pathway in KEGG (id: hsa05216) are not connected (<xref ref-type=\"fig\" rid=\"pone-0084227-g002\">Figure 2</xref>). Traditional pathway analyses consider all the molecules (genes) in the same pathway together. We believe that the non-interacting set of molecules should be considered separately. Therefore, we propose to include only the interacting molecules in a candidate set, which we call the <italic>Molecular Interaction Subnetwork (MIS)</italic>.</p>\n",
       "<fig id=\"pone-0084227-g002\" orientation=\"portrait\" position=\"float\">\n",
       "<object-id pub-id-type=\"doi\">10.1371/journal.pone.0084227.g002</object-id>\n",
       "<label>Figure 2</label>\n",
       "<caption>\n",
       "<title>The thyroid cancer pathway in KEGG (ID: hsa05216).</title>\n",
       "<p>The top rectangle marked by C1 shows that REDPTC and TRK have an indirect effect on the activation of Ras. Ras then activates BRAF, and BRAF phosphorylates MEK which in turn phosphorylates ERK. The result of this path is proliferation survival. This pathway has five connected components (C1–C5). Among them, C2 is actually a subset of C1, and the others are fully disconnected, i.e., there is no significant interaction between any pair of the components.</p>\n",
       "</caption>\n",
       "<graphic xlink:href=\"pone.0084227.g002\"></graphic>\n",
       "</fig>\n",
       "<p>We further adjust the idea of an MIS for large number vertices in a connected network. Some pathways such as the cell cycle pathway in KEGG (id: hsa04110) contain a large number of genes all of which are connected to each other, although different parts of the pathway are annotated to have different functions. We believe that, it would not be wise to assume that each and every molecule in a large network responsible for a process would also be responsible for another process. Instead, we reasonably assume that a subset of molecules in one process may also affect the development of another process. Hence, we propose to divide the large network into pieces so that only the closely interacting molecules are in the same subnetwork. An MIS will contain all the molecules in such a subnetwork.</p>\n",
       "<p>We utilize the community structure of a network –groups of nodes within which the edge densities are high, but between which they are low – to find the closely interacting subnetworks. We first build a community dendrogram as described by <xref ref-type=\"bibr\" rid=\"pone.0084227-Geman1\">[4]</xref>. Smaller communities can be made by splitting the dendrogram from the top. Modularity, a measure of the quality of a particular division of a network, is based on the ratio of the number of edges within groups to the total number of edges in the network. The division improves as the modularity increases. Among all the available ones, the pair of groups of nodes whose integration would produce the highest modularity are joined until one group remains. This process can be presented as a dendrogram whose division from the top to the bottom would create smaller communities. If the community size is within an <italic>appropriate range [minRange, maxRange]</italic>, we consider it as an <italic>appropriate community</italic>. If the size is above the appropriate range, we cut the community dendrogram from the top. So we get two small communities and we recursively do this until the size is within the appropriate range. The partition process discards any branch whose size goes below the appropriate range and once the partition process is over, each node in the discarded branch is individually merged with the closest appropriate community. This merging step may cause some communities to grow above the appropriate range. For each of these communities, we create a new network taking all the nodes of the community and the edges among them from the original network, and then we apply the whole process to the new network. Algorithm 1, along with Algorithm 2–4, describes the MIS generation process in detail. The algorithm is illustrated in Figure S1 in Supporting <xref ref-type=\"supplementary-material\" rid=\"pone.0084227.s001\">Document S1</xref>.</p>\n",
       "<boxed-text orientation=\"portrait\" position=\"float\">\n",
       "<sec id=\"s1b1a\">\n",
       "<title>Algorithm 1 Generate MIS from network</title>\n",
       "<p>\n",
       "<bold>Input:</bold>\n",
       "<inline-formula><inline-graphic xlink:href=\"pone.0084227.e030.jpg\"></inline-graphic></inline-formula>, a molecular interaction network</p>\n",
       "<p>\n",
       "<bold>Output:</bold> List of MISs generated from <inline-formula><inline-graphic xlink:href=\"pone.0084227.e031.jpg\"></inline-graphic></inline-formula>\n",
       "</p>\n",
       "<p>1: <bold>for all</bold> Connected component <inline-formula><inline-graphic xlink:href=\"pone.0084227.e032.jpg\"></inline-graphic></inline-formula> in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e033.jpg\"></inline-graphic></inline-formula>\n",
       "<bold>do</bold>\n",
       "</p>\n",
       "<p>2: <bold>if</bold>\n",
       "<inline-formula><inline-graphic xlink:href=\"pone.0084227.e034.jpg\"></inline-graphic></inline-formula>\n",
       "<bold>then</bold>\n",
       "</p>\n",
       "<p>3: Generate MISs from <inline-formula><inline-graphic xlink:href=\"pone.0084227.e035.jpg\"></inline-graphic></inline-formula> using Algorithm 2 and save them.</p>\n",
       "<p>4: <bold>eles</bold>\n",
       "</p>\n",
       "<p>5: Generate an MIS taking all the molecules in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e036.jpg\"></inline-graphic></inline-formula> and save it</p>\n",
       "<p>6: <bold>end if</bold>\n",
       "</p>\n",
       "<p>7: <bold>end for</bold>\n",
       "</p>\n",
       "<p>8: <bold>return</bold> all saved MISs</p>\n",
       "</sec>\n",
       "</boxed-text>\n",
       "<boxed-text orientation=\"portrait\" position=\"float\">\n",
       "<sec id=\"s1b2a\">\n",
       "<title>Algorithm 2 Generate MIS from connected network</title>\n",
       "<p>\n",
       "<bold>Input:</bold>\n",
       "<bold>Input:</bold>\n",
       "<inline-formula><inline-graphic xlink:href=\"pone.0084227.e037.jpg\"></inline-graphic></inline-formula>, a connected network with at least <inline-formula><inline-graphic xlink:href=\"pone.0084227.e038.jpg\"></inline-graphic></inline-formula> nodes.</p>\n",
       "<p>\n",
       "<bold>Output:</bold> MISs generated from <inline-formula><inline-graphic xlink:href=\"pone.0084227.e039.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>1: <inline-formula><inline-graphic xlink:href=\"pone.0084227.e040.jpg\"></inline-graphic></inline-formula> Build a community dendrogram as described by <xref ref-type=\"bibr\" rid=\"pone.0084227-Clauset1\">[18]</xref>.</p>\n",
       "<p>2: <inline-formula><inline-graphic xlink:href=\"pone.0084227.e041.jpg\"></inline-graphic></inline-formula> appropriate communities generated from <inline-formula><inline-graphic xlink:href=\"pone.0084227.e042.jpg\"></inline-graphic></inline-formula> using Algorithm 3.</p>\n",
       "<p>3: <inline-formula><inline-graphic xlink:href=\"pone.0084227.e043.jpg\"></inline-graphic></inline-formula> communities after merging the discarded nodes in step 2 with the closest appropriate communities using Algorithm 4.</p>\n",
       "<p>4: <bold>for all</bold> community <inline-formula><inline-graphic xlink:href=\"pone.0084227.e044.jpg\"></inline-graphic></inline-formula> in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e045.jpg\"></inline-graphic></inline-formula>\n",
       "<bold>do.</bold>\n",
       "</p>\n",
       "<p>5: <bold>if</bold> size(<inline-formula><inline-graphic xlink:href=\"pone.0084227.e046.jpg\"></inline-graphic></inline-formula>)<inline-formula><inline-graphic xlink:href=\"pone.0084227.e047.jpg\"></inline-graphic></inline-formula>\n",
       "<inline-formula><inline-graphic xlink:href=\"pone.0084227.e048.jpg\"></inline-graphic></inline-formula>\n",
       "<bold>then.</bold>\n",
       "</p>\n",
       "<p>6: Create and save an MIS taking all the molecules in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e049.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>7: <bold>else.</bold>\n",
       "</p>\n",
       "<p>8: <inline-formula><inline-graphic xlink:href=\"pone.0084227.e050.jpg\"></inline-graphic></inline-formula>\n",
       "<inline-formula><inline-graphic xlink:href=\"pone.0084227.e051.jpg\"></inline-graphic></inline-formula> a subgraph of <inline-formula><inline-graphic xlink:href=\"pone.0084227.e052.jpg\"></inline-graphic></inline-formula> containing all the leaf nodes of <inline-formula><inline-graphic xlink:href=\"pone.0084227.e053.jpg\"></inline-graphic></inline-formula> and all the edges among them in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e054.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>9: Recursively generate and save MISs from <inline-formula><inline-graphic xlink:href=\"pone.0084227.e055.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>10: <bold>end if.</bold>\n",
       "</p>\n",
       "<p>11: <bold>end for.</bold>\n",
       "</p>\n",
       "<p>12: <bold>return</bold> all saved MISs.</p>\n",
       "</sec>\n",
       "</boxed-text>\n",
       "<boxed-text orientation=\"portrait\" position=\"float\">\n",
       "<sec id=\"s1b3a\">\n",
       "<title>Algorithm 3 Extract appropriate communities from dendrogram</title>\n",
       "<p>\n",
       "<bold>Input:</bold>\n",
       "<inline-formula><inline-graphic xlink:href=\"pone.0084227.e056.jpg\"></inline-graphic></inline-formula>, the community dendrogram.</p>\n",
       "<p>\n",
       "<bold>Output:</bold> Appropriate communities extracted from <inline-formula><inline-graphic xlink:href=\"pone.0084227.e057.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>1: <inline-formula><inline-graphic xlink:href=\"pone.0084227.e058.jpg\"></inline-graphic></inline-formula>\n",
       "<inline-formula><inline-graphic xlink:href=\"pone.0084227.e059.jpg\"></inline-graphic></inline-formula> No. of leaf nodes in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e060.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>2: <bold>if</bold>\n",
       "<inline-formula><inline-graphic xlink:href=\"pone.0084227.e061.jpg\"></inline-graphic></inline-formula>\n",
       "<bold>then.</bold>\n",
       "</p>\n",
       "<p>3: Create a community of all the leaf nodes in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e062.jpg\"></inline-graphic></inline-formula> and save it.</p>\n",
       "<p>4: <bold>even if</bold>\n",
       "<inline-formula><inline-graphic xlink:href=\"pone.0084227.e063.jpg\"></inline-graphic></inline-formula>\n",
       "<bold>then.</bold>\n",
       "</p>\n",
       "<p>5: Divide <inline-formula><inline-graphic xlink:href=\"pone.0084227.e064.jpg\"></inline-graphic></inline-formula> into 2 parts from the top - <inline-formula><inline-graphic xlink:href=\"pone.0084227.e065.jpg\"></inline-graphic></inline-formula> and <inline-formula><inline-graphic xlink:href=\"pone.0084227.e066.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>6: Recursively extract and save appropriate communities from <inline-formula><inline-graphic xlink:href=\"pone.0084227.e067.jpg\"></inline-graphic></inline-formula> and <inline-formula><inline-graphic xlink:href=\"pone.0084227.e068.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>7: <bold>eles.</bold>\n",
       "</p>\n",
       "<p>8: Discard the <inline-formula><inline-graphic xlink:href=\"pone.0084227.e069.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>9: <bold>end if.</bold>\n",
       "</p>\n",
       "<p>10: <bold>return</bold> saved appropriate communities.</p>\n",
       "</sec>\n",
       "</boxed-text>\n",
       "<boxed-text orientation=\"portrait\" position=\"float\">\n",
       "<sec id=\"s1b4a\">\n",
       "<title>Algorithm 4 Merge discarded nodes with given communities</title>\n",
       "<p>\n",
       "<bold>Input:</bold>\n",
       "<inline-formula><inline-graphic xlink:href=\"pone.0084227.e070.jpg\"></inline-graphic></inline-formula>, nodes not taken in any community; <inline-formula><inline-graphic xlink:href=\"pone.0084227.e071.jpg\"></inline-graphic></inline-formula>, given communities.</p>\n",
       "<p>\n",
       "<bold>Output:</bold> Communities with <inline-formula><inline-graphic xlink:href=\"pone.0084227.e072.jpg\"></inline-graphic></inline-formula> merged.</p>\n",
       "<p>1: <bold>repeat.</bold>\n",
       "</p>\n",
       "<p>2: <inline-formula><inline-graphic xlink:href=\"pone.0084227.e073.jpg\"></inline-graphic></inline-formula> The subset of <inline-formula><inline-graphic xlink:href=\"pone.0084227.e074.jpg\"></inline-graphic></inline-formula> directly connected (1-hop distance) to some node in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e075.jpg\"></inline-graphic></inline-formula> from the original network.</p>\n",
       "<p>3: <bold>for all</bold> node <inline-formula><inline-graphic xlink:href=\"pone.0084227.e076.jpg\"></inline-graphic></inline-formula> in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e077.jpg\"></inline-graphic></inline-formula>\n",
       "<bold>do.</bold>\n",
       "</p>\n",
       "<p>4: <inline-formula><inline-graphic xlink:href=\"pone.0084227.e078.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>5: <inline-formula><inline-graphic xlink:href=\"pone.0084227.e079.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>6: <bold>loop.</bold>\n",
       "</p>\n",
       "<p>7: <inline-formula><inline-graphic xlink:href=\"pone.0084227.e080.jpg\"></inline-graphic></inline-formula> Count no. of <inline-formula><inline-graphic xlink:href=\"pone.0084227.e081.jpg\"></inline-graphic></inline-formula>-hop distant nodes (from <inline-formula><inline-graphic xlink:href=\"pone.0084227.e082.jpg\"></inline-graphic></inline-formula>) in each community in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e083.jpg\"></inline-graphic></inline-formula>. <inline-formula><inline-graphic xlink:href=\"pone.0084227.e084.jpg\"></inline-graphic></inline-formula> communities having <inline-formula><inline-graphic xlink:href=\"pone.0084227.e085.jpg\"></inline-graphic></inline-formula>\n",
       "<inline-formula><inline-graphic xlink:href=\"pone.0084227.e086.jpg\"></inline-graphic></inline-formula>-hop distant nodes from <inline-formula><inline-graphic xlink:href=\"pone.0084227.e087.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>9: <bold>if</bold>\n",
       "<inline-formula><inline-graphic xlink:href=\"pone.0084227.e088.jpg\"></inline-graphic></inline-formula>  = 0 <bold>then.</bold>\n",
       "</p>\n",
       "<p>10: Mark <inline-formula><inline-graphic xlink:href=\"pone.0084227.e089.jpg\"></inline-graphic></inline-formula>'s membership to one community in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e090.jpg\"></inline-graphic></inline-formula> randomly.</p>\n",
       "<p>11: <bold>break</bold> loop.</p>\n",
       "<p>12:<bold> else if</bold>\n",
       "<inline-formula><inline-graphic xlink:href=\"pone.0084227.e091.jpg\"></inline-graphic></inline-formula>  = 1 <bold>then.</bold>\n",
       "</p>\n",
       "<p>13: Mark <inline-formula><inline-graphic xlink:href=\"pone.0084227.e092.jpg\"></inline-graphic></inline-formula>'s membership to the only community in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e093.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>14: <bold>break</bold> loop.</p>\n",
       "<p>15:<bold> else.</bold>\n",
       "</p>\n",
       "<p>16: <inline-formula><inline-graphic xlink:href=\"pone.0084227.e094.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>17: <inline-formula><inline-graphic xlink:href=\"pone.0084227.e095.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>18: <bold>end if.</bold>\n",
       "</p>\n",
       "<p>19: <bold>end loop.</bold>\n",
       "</p>\n",
       "<p>20: <bold>end for.</bold>\n",
       "</p>\n",
       "<p>21: <inline-formula><inline-graphic xlink:href=\"pone.0084227.e096.jpg\"></inline-graphic></inline-formula> communities after merging every node in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e097.jpg\"></inline-graphic></inline-formula> according to its membership.</p>\n",
       "<p>22: <inline-formula><inline-graphic xlink:href=\"pone.0084227.e098.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>23: <bold>until</bold>\n",
       "<inline-formula><inline-graphic xlink:href=\"pone.0084227.e099.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "<p>24: <bold>return</bold>\n",
       "<inline-formula><inline-graphic xlink:href=\"pone.0084227.e100.jpg\"></inline-graphic></inline-formula>.</p>\n",
       "</sec>\n",
       "</boxed-text>\n",
       "</sec>\n",
       "<sec id=\"s1c\">\n",
       "<title>Feature Selection</title>\n",
       "<p>It is necessary to map the molecules to the probes used in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e101.jpg\"></inline-graphic></inline-formula>. If one molecule is mapped to multiple probes we include all of them separately following the suggestion to consider each probe independently <xref ref-type=\"bibr\" rid=\"pone.0084227-Stalteri1\">[19]</xref>. An MIS is selected if its molecules are mapped to at least a certain number of unique probes. Thus we avoid the unreliable effect of a single or low number of probes. For every selected MIS, we propose to take the same number of probes to consider them equally in the next step. In this study, we choose the five most differentially expressed probes in each MIS which we call the <italic>representative probeset</italic> for the MIS. We estimate the differential expression of a probe using a modified version of Welch’s t-statistic score (Section S1 in <xref ref-type=\"supplementary-material\" rid=\"pone.0084227.s001\">Document S1</xref>).</p>\n",
       "<p>As one molecule can be mapped to multiple probes and can be in multiple nodes of some molecular interaction networks, the representative probesets of MISs may overlap with each other. We measure the overlap as the ratio of the number of common probes in the representative probesets of the two MISs over the number of probes in each representative probeset, <inline-formula><inline-graphic xlink:href=\"pone.0084227.e102.jpg\"></inline-graphic></inline-formula>, where RP(.) denotes the representative probeset of an MIS. We take two MISs with the highest overlap and combine them into one MIS by taking a union of all the molecules in both MISs. We continue to combine MISs with the highest overlaps as long as they are significant (<inline-formula><inline-graphic xlink:href=\"pone.0084227.e103.jpg\"></inline-graphic></inline-formula> in this study).</p>\n",
       "</sec>\n",
       "<sec id=\"s1d\">\n",
       "<title>Ranking</title>\n",
       "<p>We rank the MISs based on the expression pattern of their representative probesets. We cluster all the samples using the expression values of the representative probeset. Each cluster represents a separate pattern. If a specific pattern corresponds to a class, we can expect that almost all the samples in the corresponding cluster will be of the same class (<xref ref-type=\"fig\" rid=\"pone-0084227-g003\">Figure 3</xref>). Based on this behavior, we propose an entropy-based score to rank the MISs.</p>\n",
       "<fig id=\"pone-0084227-g003\" orientation=\"portrait\" position=\"float\">\n",
       "<object-id pub-id-type=\"doi\">10.1371/journal.pone.0084227.g003</object-id>\n",
       "<label>Figure 3</label>\n",
       "<caption>\n",
       "<title>Ranking of MIS.</title>\n",
       "<p>Let three probes (i, j, and k) constitute the representative probeset of an MIS. We plot all the samples with the expression values of these probes in separate dimensions, and then we cluster the samples. If the samples in a cluster are mostly of one kind (such as <inline-formula><inline-graphic xlink:href=\"pone.0084227.e104.jpg\"></inline-graphic></inline-formula> or <inline-formula><inline-graphic xlink:href=\"pone.0084227.e105.jpg\"></inline-graphic></inline-formula>), we can say that the cluster’s expression pattern represents the corresponding class (positive or negative). The ranking of an MIS producing such clusters should be high.</p>\n",
       "</caption>\n",
       "<graphic xlink:href=\"pone.0084227.g003\"></graphic>\n",
       "</fig>\n",
       "<p>Let us consider that all the samples in dataset <inline-formula><inline-graphic xlink:href=\"pone.0084227.e106.jpg\"></inline-graphic></inline-formula> have been divided into <inline-formula><inline-graphic xlink:href=\"pone.0084227.e107.jpg\"></inline-graphic></inline-formula> clusters <inline-formula><inline-graphic xlink:href=\"pone.0084227.e108.jpg\"></inline-graphic></inline-formula> using the expression values of all the probes in an MIS representative probeset. Each cluster <inline-formula><inline-graphic xlink:href=\"pone.0084227.e109.jpg\"></inline-graphic></inline-formula> contains <inline-formula><inline-graphic xlink:href=\"pone.0084227.e110.jpg\"></inline-graphic></inline-formula> positive samples and <inline-formula><inline-graphic xlink:href=\"pone.0084227.e111.jpg\"></inline-graphic></inline-formula> negative samples. We normalize the number of positive and negative samples in each cluster by dividing it by the total number of positive and negative samples in the dataset, respectively. The normalized number of positive samples is <inline-formula><inline-graphic xlink:href=\"pone.0084227.e112.jpg\"></inline-graphic></inline-formula> and the normalized number of negative samples is <inline-formula><inline-graphic xlink:href=\"pone.0084227.e113.jpg\"></inline-graphic></inline-formula>. The entropy of cluster <inline-formula><inline-graphic xlink:href=\"pone.0084227.e114.jpg\"></inline-graphic></inline-formula> is then defined as<disp-formula id=\"pone.0084227.e115\"><graphic orientation=\"portrait\" position=\"anchor\" xlink:href=\"pone.0084227.e115.jpg\"></graphic><label>(1)</label></disp-formula>\n",
       "</p>\n",
       "<p>The entropy of the MIS is then defined as<disp-formula id=\"pone.0084227.e116\"><graphic orientation=\"portrait\" position=\"anchor\" xlink:href=\"pone.0084227.e116.jpg\"></graphic><label>(2)</label></disp-formula>\n",
       "</p>\n",
       "<p>The range of each cluster’s entropy is <inline-formula><inline-graphic xlink:href=\"pone.0084227.e117.jpg\"></inline-graphic></inline-formula>. Consequently, the entropy of an MIS is always within <inline-formula><inline-graphic xlink:href=\"pone.0084227.e118.jpg\"></inline-graphic></inline-formula>. If all the samples in every cluster are of one kind (same class), then the entropy becomes zero. In contrast, it becomes one if every cluster contains an equal proportion of positive and negative samples. MISs are sorted by their entropies in an increasing order, i.e., the MISs with the lowest and highest entropies are at the top and bottom of the ranking, respectively.</p>\n",
       "</sec>\n",
       "<sec id=\"s1e\">\n",
       "<title>Context Prediction</title>\n",
       "<p>The context (or class) of a new sample <inline-formula><inline-graphic xlink:href=\"pone.0084227.e119.jpg\"></inline-graphic></inline-formula> is predicted from the weighted voting by the top <inline-formula><inline-graphic xlink:href=\"pone.0084227.e120.jpg\"></inline-graphic></inline-formula> MISs <inline-formula><inline-graphic xlink:href=\"pone.0084227.e121.jpg\"></inline-graphic></inline-formula>. <inline-formula><inline-graphic xlink:href=\"pone.0084227.e122.jpg\"></inline-graphic></inline-formula> votes for the positive and negative class with the weight of <inline-formula><inline-graphic xlink:href=\"pone.0084227.e123.jpg\"></inline-graphic></inline-formula> and <inline-formula><inline-graphic xlink:href=\"pone.0084227.e124.jpg\"></inline-graphic></inline-formula>, respectively (Eqs. 3 and 4).<disp-formula id=\"pone.0084227.e125\"><graphic orientation=\"portrait\" position=\"anchor\" xlink:href=\"pone.0084227.e125.jpg\"></graphic><label>(3)</label></disp-formula>\n",
       "<disp-formula id=\"pone.0084227.e126\"><graphic orientation=\"portrait\" position=\"anchor\" xlink:href=\"pone.0084227.e126.jpg\"></graphic><label>(4)</label></disp-formula>where <inline-formula><inline-graphic xlink:href=\"pone.0084227.e127.jpg\"></inline-graphic></inline-formula> is the closest cluster to <inline-formula><inline-graphic xlink:href=\"pone.0084227.e128.jpg\"></inline-graphic></inline-formula> among the clusters produced by <inline-formula><inline-graphic xlink:href=\"pone.0084227.e129.jpg\"></inline-graphic></inline-formula>, and <inline-formula><inline-graphic xlink:href=\"pone.0084227.e130.jpg\"></inline-graphic></inline-formula> and <inline-formula><inline-graphic xlink:href=\"pone.0084227.e131.jpg\"></inline-graphic></inline-formula> are the normalized numbers of the positive and negative samples, respectively, in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e132.jpg\"></inline-graphic></inline-formula>. The sums of the voting weights by the top <inline-formula><inline-graphic xlink:href=\"pone.0084227.e133.jpg\"></inline-graphic></inline-formula> MISs for the positive class and the negative class are <inline-formula><inline-graphic xlink:href=\"pone.0084227.e134.jpg\"></inline-graphic></inline-formula> and <inline-formula><inline-graphic xlink:href=\"pone.0084227.e135.jpg\"></inline-graphic></inline-formula>, respectively (Eqs. 5 and 6).</p>\n",
       "<p>\n",
       "<disp-formula id=\"pone.0084227.e136\"><graphic orientation=\"portrait\" position=\"anchor\" xlink:href=\"pone.0084227.e136.jpg\"></graphic><label>(5)</label></disp-formula>\n",
       "<disp-formula id=\"pone.0084227.e137\"><graphic orientation=\"portrait\" position=\"anchor\" xlink:href=\"pone.0084227.e137.jpg\"></graphic><label>(6)</label></disp-formula>\n",
       "</p>\n",
       "<p>The class, with a voting weight that is higher than the other, is predicted to be the class of the new sample, <inline-formula><inline-graphic xlink:href=\"pone.0084227.e138.jpg\"></inline-graphic></inline-formula>. However, if both weights, <inline-formula><inline-graphic xlink:href=\"pone.0084227.e139.jpg\"></inline-graphic></inline-formula> and <inline-formula><inline-graphic xlink:href=\"pone.0084227.e140.jpg\"></inline-graphic></inline-formula>, become equal, the class is predicted from <italic>binary voting</italic>: a class gets the full weight (i.e., one) if its number of samples surpasses that of the other class in a nearby cluster; otherwise it gets zero (Section S2 in Supporting <xref ref-type=\"supplementary-material\" rid=\"pone.0084227.s001\">Document S1</xref>). Binary voting would take place infrequently as we use only the top ranked MISs. In fact, it did not occur in any of our experiments. The binary voting system guarantees the winner when <inline-formula><inline-graphic xlink:href=\"pone.0084227.e141.jpg\"></inline-graphic></inline-formula> is odd. The classification rule is finally expressed as below (Eq. 7).<disp-formula id=\"pone.0084227.e142\"><graphic orientation=\"portrait\" position=\"anchor\" xlink:href=\"pone.0084227.e142.jpg\"></graphic><label>(7)</label></disp-formula>\n",
       "</p>\n",
       "</sec>\n",
       "</sec>\n",
       "<sec id=\"s2\">\n",
       "<title>Datasets, Implementation and Evaluation Method</title>\n",
       "<sec id=\"s2a\">\n",
       "<title>Microarray Datasets</title>\n",
       "<p>We tested COSSY on seven publicly available cancer datasets, as shown in <xref ref-type=\"table\" rid=\"pone-0084227-t001\">Table 1</xref>. In this paper, we refer to the datasets by the names mentioned in the first column of <xref ref-type=\"table\" rid=\"pone-0084227-t001\">Table 1</xref>. The purpose of the Leukemia microarray dataset is to identify genes that distinguish Acute Myeloid Leukemia (AML), a deathly disease, from Acute Lymphoblastic Leukemia (ALL) <xref ref-type=\"bibr\" rid=\"pone.0084227-Golub1\">[20]</xref>. It contains expression levels of 7129 probes from 72 samples (ALL:47, AML:25). We used dataset <italic>B</italic> from <xref ref-type=\"bibr\" rid=\"pone.0084227-Pomeroy1\">[21]</xref>, which contains 25 classic and 9 demoplastic medulloblastoma samples, as the CNS dataset. The samples of the DLBCL dataset are divided into 2 categories: Diffuse Large B-Cell Lymphoma (DLBCL) and Follicular Lymphoma (FL) <xref ref-type=\"bibr\" rid=\"pone.0084227-Shipp1\">[22]</xref>. The Prostate1 dataset, distinguishing between prostate tumors and non-tumor prostates (normal), used 102 high-quality expression profiles from <xref ref-type=\"bibr\" rid=\"pone.0084227-Singh1\">[23]</xref>. The Prostate3 dataset includes 24 prostate tumor and 9 normal samples from <xref ref-type=\"bibr\" rid=\"pone.0084227-Welsh1\">[24]</xref>. The goal of the Lung dataset is to differentiate between Malignant Pleural Mesothelioma (MPM) and Adenocarcinoma (ADCA) of the lung <xref ref-type=\"bibr\" rid=\"pone.0084227-Gordon1\">[25]</xref>. Finally, the <italic>GCM_total</italic> dataset from <xref ref-type=\"bibr\" rid=\"pone.0084227-Ramaswamy1\">[26]</xref>, which contains 90 normal and 190 tumor samples of several types of cancer, constitutes the GCM dataset. The sources of these datasets have been listed in Table S1 in Supporting <xref ref-type=\"supplementary-material\" rid=\"pone.0084227.s001\">Document S1</xref>.</p>\n",
       "<table-wrap id=\"pone-0084227-t001\" orientation=\"portrait\" position=\"float\">\n",
       "<object-id pub-id-type=\"doi\">10.1371/journal.pone.0084227.t001</object-id>\n",
       "<label>Table 1</label>\n",
       "<caption>\n",
       "<title>Microarray Datasets.</title>\n",
       "</caption>\n",
       "<alternatives>\n",
       "<graphic id=\"pone-0084227-t001-1\" xlink:href=\"pone.0084227.t001\"></graphic>\n",
       "<table frame=\"hsides\" rules=\"groups\">\n",
       "<colgroup span=\"1\">\n",
       "<col align=\"left\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "</colgroup>\n",
       "<thead>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Dataset Name</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">#Probes</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Positive Class (#samples)</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Negative Class (#samples)</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Reference</td>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Leukemia</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">7129</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">AML (25)</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">ALL (47)</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<xref ref-type=\"bibr\" rid=\"pone.0084227-Golub1\">[20]</xref>\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">CNS</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">7129</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Demoplastic (9)</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Classic (25)</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<xref ref-type=\"bibr\" rid=\"pone.0084227-Pomeroy1\">[21]</xref>\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">DLBCL</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">7129</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">DLBCL (58)</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">FL (19)</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<xref ref-type=\"bibr\" rid=\"pone.0084227-Shipp1\">[22]</xref>\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Prostate1</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">12600</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Tumor (52)</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Normal (50)</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<xref ref-type=\"bibr\" rid=\"pone.0084227-Singh1\">[23]</xref>\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Prostate3</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">12626</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Tumor (24)</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Normal (9)</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<xref ref-type=\"bibr\" rid=\"pone.0084227-Welsh1\">[24]</xref>\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Lung</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">12533</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">MPM (31)</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">ADCA (150)</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<xref ref-type=\"bibr\" rid=\"pone.0084227-Gordon1\">[25]</xref>\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">GCM</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">16063</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Tumor (190)</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Normal (90)</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<xref ref-type=\"bibr\" rid=\"pone.0084227-Ramaswamy1\">[26]</xref>\n",
       "</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</alternatives>\n",
       "<table-wrap-foot>\n",
       "<fn id=\"nt101\">\n",
       "<label></label>\n",
       "<p>The first column, ‘Dataset Name’, indicates the name of the microarray dataset used in the manuscript. ‘#Probes’ shows the number of probes present in the dataset. The third and fourth columns contain the name of the positive and negative class, respectively, followed by the number of samples of that class. The last column shows the reference the dataset was collected from.</p>\n",
       "</fn>\n",
       "</table-wrap-foot>\n",
       "</table-wrap>\n",
       "</sec>\n",
       "<sec id=\"s2b\">\n",
       "<title>Implementation of COSSY</title>\n",
       "<p>COSSY is implemented as an R package. It requires R (<inline-formula><inline-graphic xlink:href=\"pone.0084227.e143.jpg\"></inline-graphic></inline-formula>2.15.3) along with utils, stats, limma, and igraph (<inline-formula><inline-graphic xlink:href=\"pone.0084227.e144.jpg\"></inline-graphic></inline-formula>0.6.5) packages. COSSY has been tested on Windows, Linux, and Mac operating systems. The package and data used in this study are available for download on our website: <ext-link ext-link-type=\"uri\" xlink:href=\"http://infos.korea.ac.kr/cossy/\">http://infos.korea.ac.kr/cossy/</ext-link>.</p>\n",
       "</sec>\n",
       "<sec id=\"s2c\">\n",
       "<title>Experimental Settings</title>\n",
       "<p>We first performed the quantile normalization over the training dataset and then standardized the probe data using z-score, <inline-formula><inline-graphic xlink:href=\"pone.0084227.e145.jpg\"></inline-graphic></inline-formula>, where <inline-formula><inline-graphic xlink:href=\"pone.0084227.e146.jpg\"></inline-graphic></inline-formula> is the expression value of the <inline-formula><inline-graphic xlink:href=\"pone.0084227.e147.jpg\"></inline-graphic></inline-formula>-th probe in the <inline-formula><inline-graphic xlink:href=\"pone.0084227.e148.jpg\"></inline-graphic></inline-formula>-th sample; <inline-formula><inline-graphic xlink:href=\"pone.0084227.e149.jpg\"></inline-graphic></inline-formula> and <inline-formula><inline-graphic xlink:href=\"pone.0084227.e150.jpg\"></inline-graphic></inline-formula> are the mean and standard deviation of the expression values, respectively, of the <inline-formula><inline-graphic xlink:href=\"pone.0084227.e151.jpg\"></inline-graphic></inline-formula>-th probe. Ward’s minimum variance agglomerative hierarchical clustering <xref ref-type=\"bibr\" rid=\"pone.0084227-Ward1\">[27]</xref> was used with euclidean distance. The dendrogram produced by Ward’s method was cut into <inline-formula><inline-graphic xlink:href=\"pone.0084227.e152.jpg\"></inline-graphic></inline-formula> groups to form <inline-formula><inline-graphic xlink:href=\"pone.0084227.e153.jpg\"></inline-graphic></inline-formula> clusters. The number of clusters, <inline-formula><inline-graphic xlink:href=\"pone.0084227.e154.jpg\"></inline-graphic></inline-formula>, was chosen from the <italic>rule of thumb</italic>, <inline-formula><inline-graphic xlink:href=\"pone.0084227.e155.jpg\"></inline-graphic></inline-formula>, where <inline-formula><inline-graphic xlink:href=\"pone.0084227.e156.jpg\"></inline-graphic></inline-formula> is the number of samples <xref ref-type=\"bibr\" rid=\"pone.0084227-Mardia1\">[28]</xref>. To avoid the effect of noisy data, the median of the data points in each cluster was taken as its centroid. A new data point would fall in the cluster whose centroid is the closest to the new point. Finally, <inline-formula><inline-graphic xlink:href=\"pone.0084227.e157.jpg\"></inline-graphic></inline-formula>, the number of the top MISs to vote, was selected experimentally. We experimented with <inline-formula><inline-graphic xlink:href=\"pone.0084227.e158.jpg\"></inline-graphic></inline-formula> and used the minimum <inline-formula><inline-graphic xlink:href=\"pone.0084227.e159.jpg\"></inline-graphic></inline-formula> that produced the highest accuracy.</p>\n",
       "</sec>\n",
       "<sec id=\"s2d\">\n",
       "<title>Cross Validation</title>\n",
       "<p>We validate the classification of COSSY by standard <italic>Leave-One-Out Cross-Validation (LOOCV)</italic>. For each sample <inline-formula><inline-graphic xlink:href=\"pone.0084227.e160.jpg\"></inline-graphic></inline-formula> in <inline-formula><inline-graphic xlink:href=\"pone.0084227.e161.jpg\"></inline-graphic></inline-formula>, we train the classifier using the remaining (<inline-formula><inline-graphic xlink:href=\"pone.0084227.e162.jpg\"></inline-graphic></inline-formula>) samples and then the classifier predicts the class of <inline-formula><inline-graphic xlink:href=\"pone.0084227.e163.jpg\"></inline-graphic></inline-formula>. The LOOCV accuracy is the fraction of all the samples that are correctly classified.</p>\n",
       "</sec>\n",
       "</sec>\n",
       "<sec id=\"s3\">\n",
       "<title>Results</title>\n",
       "<sec id=\"s3a\">\n",
       "<title>Networks Generated from KEGG and STRING</title>\n",
       "<p>We applied COSSY on two well-known molecular interaction networks: KEGG and STRING. We chose KEGG and STRING, as they are both very different in nature. KEGG is composed of a number of disjoint pathways where the same molecule can be present in multiple pathways. In contrast, a molecule in STRING can have only one node. The nodes in KEGG may represent different types of objects (gene, gene product, chemical compound, etc.) while the nodes in STRING represent proteins. A KEGG node may have multiple molecules, while a STRING node always represents only one protein. The topologies of these two networks are also significantly different. STRING shows a much higher clustering tendency than KEGG. In this study, we used the unweighted and undirected graph of the <italic>protein networks</italic> – the network of the genes and their interactions – from the human signaling pathways in KEGG (release 62.0), and the human protein-protein interactions – both experimental and predicted with high confidence (<italic>score</italic>&gt;<italic>0.7</italic>) – in STRING (release 9.05). A few properties of the networks are shown in <xref ref-type=\"table\" rid=\"pone-0084227-t002\">Table 2</xref>.</p>\n",
       "<table-wrap id=\"pone-0084227-t002\" orientation=\"portrait\" position=\"float\">\n",
       "<object-id pub-id-type=\"doi\">10.1371/journal.pone.0084227.t002</object-id>\n",
       "<label>Table 2</label>\n",
       "<caption>\n",
       "<title>Network Properties of KEGG and STRING.</title>\n",
       "</caption>\n",
       "<alternatives>\n",
       "<graphic id=\"pone-0084227-t002-2\" xlink:href=\"pone.0084227.t002\"></graphic>\n",
       "<table frame=\"hsides\" rules=\"groups\">\n",
       "<colgroup span=\"1\">\n",
       "<col align=\"left\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "</colgroup>\n",
       "<thead>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Network</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">TotalNodes</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Gene (Protein) Nodes</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Total Edges</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">ConnectedComp.</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Avg. Node Degree</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Max Node Degree</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">ClusteringCoefficient</td>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">KEGG</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">19568</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">10691</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">10728</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">4494</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">1.84</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">43</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">0.19</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">STRING</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">14250</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">14250</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">215800</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">182</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">30.30</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">1110</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">0.61</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</alternatives>\n",
       "<table-wrap-foot>\n",
       "<fn id=\"nt102\">\n",
       "<label></label>\n",
       "<p>The ‘Total Nodes’ column contains the total number of nodes available in the network while the ‘Gene (Protein) Nodes’ column shows the number of nodes with at least one gene in KEGG (or one protein in STIRING). The fourth and fifth columns contain the total number of edges, and the number of connected components having at least one gene (or protein), respectively. ‘Avg. Node Degree’ represents the number of edges a node has on average. ‘Max Node Degree’ denotes the maximum number of edges a node has in the network. ‘Clustering Coefficient’ is the ratio of the triangles to the connected triples in a graph.</p>\n",
       "</fn>\n",
       "</table-wrap-foot>\n",
       "</table-wrap>\n",
       "<p>As described before, we try to keep the size of the MISs within the <italic>appropriate range</italic>. We experimented with different ranges (Table S2 and S3 in Supporting <xref ref-type=\"supplementary-material\" rid=\"pone.0084227.s001\">Document S1</xref>) and chose the range with the highest accuracy: 5–15 for KEGG and 5–25 for STRING. Here we should mention that the sizes of some MISs may be below or above the range (<xref ref-type=\"table\" rid=\"pone-0084227-t003\">Table 3</xref>). We did not discard the smaller MISs in the MIS generation step, as their number of probes may be sufficient. Also, the bigger MISs could not be further divided because of their special topologies, e.g., a star network or a highly dense network (Figure S2 in Supporting <xref ref-type=\"supplementary-material\" rid=\"pone.0084227.s001\">Document S1</xref>). We used bioDBnet <xref ref-type=\"bibr\" rid=\"pone.0084227-Mudunuri1\">[29]</xref> to map the genes or proteins to the microarray probes.</p>\n",
       "<table-wrap id=\"pone-0084227-t003\" orientation=\"portrait\" position=\"float\">\n",
       "<object-id pub-id-type=\"doi\">10.1371/journal.pone.0084227.t003</object-id>\n",
       "<label>Table 3</label>\n",
       "<caption>\n",
       "<title>Molecular Interaction Subnetwork Size.</title>\n",
       "</caption>\n",
       "<alternatives>\n",
       "<graphic id=\"pone-0084227-t003-3\" xlink:href=\"pone.0084227.t003\"></graphic>\n",
       "<table frame=\"hsides\" rules=\"groups\">\n",
       "<colgroup span=\"1\">\n",
       "<col align=\"left\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "</colgroup>\n",
       "<thead>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Network</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Appropriate Range</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">MIS below the range</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">MIS within the range</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">MIS above the range</td>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">KEGG</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">5–15</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">1925</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">629</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">9</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">STRING</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">5–25</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">170</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">847</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">23</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</alternatives>\n",
       "<table-wrap-foot>\n",
       "<fn id=\"nt103\">\n",
       "<label></label>\n",
       "<p>The table shows the number of MISs with a total of nodes below, within, and above the appropriate range.</p>\n",
       "</fn>\n",
       "</table-wrap-foot>\n",
       "</table-wrap>\n",
       "</sec>\n",
       "<sec id=\"s3b\">\n",
       "<title>Features Selected from Each Dataset</title>\n",
       "<p>The representative probesets selected from each dataset using KEGG and STRING are shown in Table S4 and S5 in Supporting <xref ref-type=\"supplementary-material\" rid=\"pone.0084227.s001\">Document S1</xref>, respectively. <xref ref-type=\"fig\" rid=\"pone-0084227-g004\">Figure 4</xref> demonstrates the top ranked KEGG MIS in the Leukemia dataset containing two types of samples: Acute Myeloid Leukemia (AML), and Acute Lymphoblastic Leukemia (ALL). The MIS was taken from three overlapped subnetworks from three different pathways: 1) Focal Adhesion Pathway (id: hsa04510), 2) Adherens Junction Pathway (id: hsa04520), and 3) Bacterial Invasion of Epithelial Cells Pathway (id: hsa05100). One probe (affymetrix id: X95735_at, gene symbol: ZYX) in its representative probeset is over-expressed in AML, while the others are under-expressed.</p>\n",
       "<fig id=\"pone-0084227-g004\" orientation=\"portrait\" position=\"float\">\n",
       "<object-id pub-id-type=\"doi\">10.1371/journal.pone.0084227.g004</object-id>\n",
       "<label>Figure 4</label>\n",
       "<caption>\n",
       "<title>The top ranked KEGG MIS in the Leukemia dataset.</title>\n",
       "<p>A–C) Three overlapped subnetworks from three different pathways constitute the MIS. D) The merged MIS is shown here. E) The expression heatmap of the representative probeset of the MIS is shown here.</p>\n",
       "</caption>\n",
       "<graphic xlink:href=\"pone.0084227.g004\"></graphic>\n",
       "</fig>\n",
       "</sec>\n",
       "<sec id=\"s3c\">\n",
       "<title>Classification Accuracy</title>\n",
       "<p>The LOOCV accuracies on the seven cancer datasets are shown in <xref ref-type=\"table\" rid=\"pone-0084227-t004\">Table 4</xref>. COSSY achieved LOOCV accuracies of 93.2% and 92.7% using KEGG and STRING, respectively, which is comparable to the state-of-the-art classifiers, <italic>k</italic>-TSP and SVM (Support Vector Machine), and better than other classifiers (DIRAC, TSP, Doublet, PAM, Decision Tree, Nearest Neighbor, Nave Bayes) including a recent network-based approach, DIRAC <xref ref-type=\"bibr\" rid=\"pone.0084227-Lin1\">[7]</xref>. <xref ref-type=\"fig\" rid=\"pone-0084227-g005\">Figure 5</xref> shows the classification accuracies of five classification methods: COSSY using KEGG, COSSY using STRING, <italic>k</italic>-TSP, SVM, and DIRAC. From this figure, it is evident that COSSY’s performance on the five datasets (Leukemia, DLBCL, Prostate1, Prostate3, and Lung) is almost equal to that of <italic>k</italic>-TSP and SVM; <italic>k</italic>-TSP outperforms others using CNS, and SVM outperforms others using GCM significantly. Notably, COSSY’s performance on six datasets is significantly better than that of DIRAC; COSSY’s performance on the remaining dataset (Prostate3) is equal to that of DIRAC. On average, COSSY’s accuracy is more than 10% higher than that of DIRAC (COSSY using KEGG - 93.2%, COSSY using STRING - 92.7%, DIRAC - 82.5%). We also report the Area Under Curve (AUC) values of Receiver Operating Characteristic (ROC) curves achieved in the cross validation by COSSY and DIRAC in <xref ref-type=\"table\" rid=\"pone-0084227-t005\">Table 5</xref>. COSSY outperforms DIRAC in terms of AUC (COSSY using KEGG - 0.948, COSSY using STRING - 0.945, DIRAC - 0.813). Even though a random classifier, which predicts the majority class, may have high accuracies due to the imbalance in the number of positive and negative samples in the datasets, the random classifier is unlikely to achieve as high AUCs as COSSY achieved.</p>\n",
       "<fig id=\"pone-0084227-g005\" orientation=\"portrait\" position=\"float\">\n",
       "<object-id pub-id-type=\"doi\">10.1371/journal.pone.0084227.g005</object-id>\n",
       "<label>Figure 5</label>\n",
       "<caption>\n",
       "<title>LOOCV accuracy of five notable classifiers.</title>\n",
       "<p>COSSY [KEGG] and COSSY [STRING] stand for COSSY using KEGG and STRING, respectively. <italic>k</italic>-TSP and DIRAC are the classification algorithms described in <xref ref-type=\"bibr\" rid=\"pone.0084227-Tan1\">[5]</xref> and <xref ref-type=\"bibr\" rid=\"pone.0084227-Eddy1\">[17]</xref>, respectively. SVM stands for the Support Vector Machine algorithm.</p>\n",
       "</caption>\n",
       "<graphic xlink:href=\"pone.0084227.g005\"></graphic>\n",
       "</fig>\n",
       "<table-wrap id=\"pone-0084227-t004\" orientation=\"portrait\" position=\"float\">\n",
       "<object-id pub-id-type=\"doi\">10.1371/journal.pone.0084227.t004</object-id>\n",
       "<label>Table 4</label>\n",
       "<caption>\n",
       "<title>LOOCV accuracy (%) of classifiers.</title>\n",
       "</caption>\n",
       "<alternatives>\n",
       "<graphic id=\"pone-0084227-t004-4\" xlink:href=\"pone.0084227.t004\"></graphic>\n",
       "<table frame=\"hsides\" rules=\"groups\">\n",
       "<colgroup span=\"1\">\n",
       "<col align=\"left\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "</colgroup>\n",
       "<thead>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Method</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Leukemia</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">CNS</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">DLBCL</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Prostate1</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Prostate3</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Lung</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">GCM</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Average</td>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">COSSY [KEGG]</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">98.6</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">85.3</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">93.5</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">90.2</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">100.0</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">99.5</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">85.0</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">93.2</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">COSSY [STRING]</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">95.8</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">88.2</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">94.8</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">90.2</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">97.0</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">98.3</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">84.6</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">92.7</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">DIRAC</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">94.8</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">72.3</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">73.4</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">62.9</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">100.0</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">98.8</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">75.2</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">82.5</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<italic>k</italic>-TSP<xref ref-type=\"table-fn\" rid=\"nt105\">*</xref>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">95.8</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">97.1</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">97.4</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">91.2</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">97.0</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">98.9</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">85.4</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">94.7</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">TSP<xref ref-type=\"table-fn\" rid=\"nt105\">*</xref>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">93.8</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">77.9</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">98.1</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">95.1</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">97.0</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">98.3</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">75.4</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">90.8</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">SVM<xref ref-type=\"table-fn\" rid=\"nt105\">*</xref>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">98.6</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">82.4</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">97.4</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">91.2</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">100.0</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">99.5</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">93.2</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">94.6</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Doublet [Sign-DT]<xref ref-type=\"table-fn\" rid=\"nt106\">+</xref>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">93.1</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">82.4</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">97.4</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">86.3</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">97.0</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">98.3</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">85.0</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">91.3</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Doublet [Sumdiff-DT]<xref ref-type=\"table-fn\" rid=\"nt106\">+</xref>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">91.7</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">70.6</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">97.4</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">82.4</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">87.9</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">95.0</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">81.4</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">86.6</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Doublet [Mul-DT]<xref ref-type=\"table-fn\" rid=\"nt106\">+</xref>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">84.7</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">55.9</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">97.4</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">86.3</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">90.9</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">92.3</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">83.2</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">84.4</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Decision Tree (DT)<xref ref-type=\"table-fn\" rid=\"nt105\">*</xref>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">73.6</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">67.7</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">80.5</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">87.3</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">84.9</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">96.1</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">77.9</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">81.1</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Nave Bayes<xref ref-type=\"table-fn\" rid=\"nt105\">*</xref>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">100.0</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">82.4</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">80.5</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">62.8</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">90.9</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">97.8</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">84.3</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">85.5</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<italic>k</italic> Nearest Neighbor<xref ref-type=\"table-fn\" rid=\"nt105\">*</xref>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">84.7</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">76.5</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">84.4</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">76.5</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">87.9</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">98.3</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">82.9</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">84.5</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">PAM<xref ref-type=\"table-fn\" rid=\"nt105\">*</xref>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">97.2</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">82.4</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">85.7</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">91.2</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">100.0</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">99.5</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">79.3</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">90.7</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</alternatives>\n",
       "<table-wrap-foot>\n",
       "<fn id=\"nt104\">\n",
       "<label></label>\n",
       "<p>The leftmost column contains the names of the methods; the rightmost column shows the average accuracy of each method for seven datasets, and other columns show the accuracy (%) for individual datasets. ‘COSSY [KEGG]’ and ‘COSSY [STRING]’ represent COSSY using KEGG and STRING, respectively. ‘DIRAC’ is the algorithm proposed in <xref ref-type=\"bibr\" rid=\"pone.0084227-Eddy1\">[17]</xref> whose LOOCV accuracies have been calculated using the matlab code published with the paper. <italic>k</italic>-TSP and TSP denote the classification algorithms described in <xref ref-type=\"bibr\" rid=\"pone.0084227-Tan1\">[5]</xref> and <xref ref-type=\"bibr\" rid=\"pone.0084227-Geman1\">[4]</xref>, respectively. SVM stands for Support Vector Machine. ‘Doublet [Sign-DT]’, ‘Doublet [Sumdiff-DT]’, and ‘Doublet [Mul-DT]’ denote the classification methods using Sign-Doublet, Sumdiff-Doublet, and Mul-Doublet, respectively, with decision trees as described in <xref ref-type=\"bibr\" rid=\"pone.0084227-Chopra1\">[6]</xref>. The last three rows contain the loocv accuracies using Nave Bayes, <italic>k</italic> Nearest Neighbor, and PAM classifier, respectively.</p>\n",
       "</fn>\n",
       "<fn id=\"nt105\">\n",
       "<label></label>\n",
       "<p>Results obtained from <xref ref-type=\"bibr\" rid=\"pone.0084227-Tan1\">[5]</xref>.</p>\n",
       "</fn>\n",
       "<fn id=\"nt106\">\n",
       "<label></label>\n",
       "<p>Results obtained from <xref ref-type=\"bibr\" rid=\"pone.0084227-Chopra1\">[6]</xref>.</p>\n",
       "</fn>\n",
       "</table-wrap-foot>\n",
       "</table-wrap>\n",
       "<table-wrap id=\"pone-0084227-t005\" orientation=\"portrait\" position=\"float\">\n",
       "<object-id pub-id-type=\"doi\">10.1371/journal.pone.0084227.t005</object-id>\n",
       "<label>Table 5</label>\n",
       "<caption>\n",
       "<title>Area Under Curve (AUC) values of COSSY and DIRAC for different datasets.</title>\n",
       "</caption>\n",
       "<alternatives>\n",
       "<graphic id=\"pone-0084227-t005-5\" xlink:href=\"pone.0084227.t005\"></graphic>\n",
       "<table frame=\"hsides\" rules=\"groups\">\n",
       "<colgroup span=\"1\">\n",
       "<col align=\"left\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "<col align=\"center\" span=\"1\"/>\n",
       "</colgroup>\n",
       "<thead>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Dataset Name</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">AUC of COSSY using KEGG</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">AUC of COSSY using STRING</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">AUC of DIRAC</td>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Leukemia</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<bold>0.986</bold>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">0.985</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">0.948</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">CNS</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">0.862</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<bold>0.876</bold>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">0.726</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">DLBCL</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<bold>0.985</bold>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">0.976</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">0.636</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Prostate1</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">0.909</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<bold>0.918</bold>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">0.635</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Prostate3</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<bold>1.000</bold>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">0.972</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<bold>1.000</bold>\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Lung</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<bold>0.999</bold>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<bold>0.999</bold>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">0.990</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">GCM</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<bold>0.896</bold>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">0.889</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">0.757</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">Average</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">\n",
       "<bold>0.948</bold>\n",
       "</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">0.945</td>\n",
       "<td align=\"left\" colspan=\"1\" rowspan=\"1\">0.813</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</alternatives>\n",
       "<table-wrap-foot>\n",
       "<fn id=\"nt107\">\n",
       "<label></label>\n",
       "<p>AUC has been calculated using the ROCR package in R <xref ref-type=\"bibr\" rid=\"pone.0084227-Sing1\">[35]</xref>. The best AUC for each dataset is highlighted in bold face.</p>\n",
       "</fn>\n",
       "</table-wrap-foot>\n",
       "</table-wrap>\n",
       "</sec>\n",
       "<sec id=\"s3d\">\n",
       "<title>Interpretation of the COSSY Results on the Leukemia Dataset</title>\n",
       "<p>As we discussed in this paper, the advantage of COSSY is its ability to automatically discover the important subnetworks to differentiate between two phenotypes (context-specific). We used the Leukemia dataset to illustrate this interpretation of the subnetworks identified by COSSY. As demonstrated in the results section, we selected the top five differentially expressed genes from each subnetwork for classification. Not surprisingly, the top subnetwork identified by COSSY contains ZYX (zyxin), which has been previously identified as the top differentially expressed gene to distinguish AML from ALL in this dataset <xref ref-type=\"bibr\" rid=\"pone.0084227-Golub1\">[20]</xref>. Other genes previously identified as gene signatures to distinguish AML from ALL include CTSD, LYN, MYB <xref ref-type=\"bibr\" rid=\"pone.0084227-Golub1\">[20]</xref>.</p>\n",
       "<p>Examining the over-expressed genes in AML across the top 15 subnetworks (Table S4 in Supporting <xref ref-type=\"supplementary-material\" rid=\"pone.0084227.s001\">Document S1</xref>), COSSY identified LYN/PI3K/AKT signaling (LYN, PIK3R2, AKT1), lysosome complex (LAMP2, CTSD, ATP6AP1), and integrin signaling pathways (ITGAX, ITGB2, FCER1G) as key pathways in driving this disease. By querying the core genes identified from the top MISs to the KEGG pathways, we could ‘stitch’ these subnetworks (MISs) together as illustrated in <xref ref-type=\"fig\" rid=\"pone-0084227-g006\">Figure 6</xref>. This ‘stitched’ network revealed three key pathways that have been described to be deregulated in AML. In particular, LYN, an Src-family kinase member, has been found to be over-expressed in AML <xref ref-type=\"bibr\" rid=\"pone.0084227-Golub1\">[20]</xref>, <xref ref-type=\"bibr\" rid=\"pone.0084227-DosSantos1\">[30]</xref>. Previous studies described LYN to be linked to mTOR <xref ref-type=\"bibr\" rid=\"pone.0084227-DosSantos1\">[30]</xref>; we revealed this link between LYN and mTOR via PIK3R2 and AKT1 by stitching these subnetworks together in this analysis.</p>\n",
       "<fig id=\"pone-0084227-g006\" orientation=\"portrait\" position=\"float\">\n",
       "<object-id pub-id-type=\"doi\">10.1371/journal.pone.0084227.g006</object-id>\n",
       "<label>Figure 6</label>\n",
       "<caption>\n",
       "<title>Stitching of the MISs found from the Leukemia dataset.</title>\n",
       "<p>The number at the end of the name of an MIS indicates its rank.</p>\n",
       "</caption>\n",
       "<graphic xlink:href=\"pone.0084227.g006\"></graphic>\n",
       "</fig>\n",
       "<p>PI3K/AKT signaling has been implicated as playing a critical role in AML <xref ref-type=\"bibr\" rid=\"pone.0084227-DosSantos1\">[30]</xref>–<xref ref-type=\"bibr\" rid=\"pone.0084227-Tamburini1\">[32]</xref>; also, inhibiting multiple components of this pathway might provide the therapeutic interventions to this lethal disease. From the stitched network, we identified ITGAX and ITGB2 as over-expressed in AML; these two integrin receptors interact with each other as well as with FCER1G. Recently, it has been found that the interactions of ITGAX/ITGB2 and FCER1G could signal SYK-dependent activation of the JAK/STAT pathway in AML, and this signaling axis might serve as a novel therapeutic target for AML <xref ref-type=\"bibr\" rid=\"pone.0084227-Oellerich1\">[33]</xref>.</p>\n",
       "<p>Finally, we observed that the multiple components of the lysosome (CTSD, LAMP2, and ATP6AP1) were over-expressed in AML. This is an interesting finding as a recent paper demonstrated that using mefloquine, an anti-malaria drug, could disrupt lysosome in AML and increase cancer cell killing effects <xref ref-type=\"bibr\" rid=\"pone.0084227-Sukhai1\">[34]</xref>. Furthermore, AML cells were less viable due to gene knock-down on LAMP2. This demonstrates that lysosome could be a new therapeutic target in AML <xref ref-type=\"bibr\" rid=\"pone.0084227-Sukhai1\">[34]</xref>. From this example, we highlight that COSSY can identify interesting and context-specific subnetworks from microarray gene expression datasets. By stitching the core genes together from these subnetworks, novel biology could be discovered. This represents an added value to COSSY as compared to static pathway-based analyses such as Gene Set Enrichment Analysis.</p>\n",
       "</sec>\n",
       "</sec>\n",
       "<sec id=\"s4\">\n",
       "<title>Discussion</title>\n",
       "<sec id=\"s4a\">\n",
       "<title>Gene Interaction</title>\n",
       "<p>Conventional differential expression analyses focus on single gene markers. Even the gene pairs identified by <italic>k</italic>-TSP or <italic>doublets</italic> do not necessarily imply a biochemical interaction between the genes. COSSY, in contrast, takes biological interactions under consideration ensuring a differential expression pattern of the representative probeset instead of a differential expression of a single gene. Consequently, it provides an opportunity to discover the development process of a phenotype.</p>\n",
       "</sec>\n",
       "<sec id=\"s4b\">\n",
       "<title>Network Partition</title>\n",
       "<p>Unlike GSEA and DIRAC, COSSY does not require any pre-partitioned subnetworks and is not limited by the quality of the predefined gene sets. COSSY can automatically generate the subnetworks. Importantly, these subnetworks may also be used with other types of analyses requiring pre-partitioned subnetworks, which is not yet possible from any of the existing subnetwork discovery methods.</p>\n",
       "</sec>\n",
       "<sec id=\"s4c\">\n",
       "<title>Non-greedy Approach</title>\n",
       "<p>The existing subnetwork discovery methods <xref ref-type=\"bibr\" rid=\"pone.0084227-Chuang1\">[13]</xref>–<xref ref-type=\"bibr\" rid=\"pone.0084227-Su1\">[16]</xref> usually build subnetworks using a greedy approach that starts with a small network (typically a single node) and then adds more nodes greedily until the addition improves the quality. However, it may have local-optima problems. On the other hand, as our method follows the non-greedy approach – divides the network first and then calculates its effectiveness – local-optima problems are unlikely to occur. Moreover, the applicability of the previous greedy methods often depends on the network topology, which is explained in the following section.</p>\n",
       "</sec>\n",
       "<sec id=\"s4d\">\n",
       "<title>COSSY is Independent of Network Topology</title>\n",
       "<p>Current subnetwork discovery methods are highly dependent on the network topology. The method by Chuang et al. (2007) <xref ref-type=\"bibr\" rid=\"pone.0084227-Chuang1\">[13]</xref> is likely to stop at a locally optimum node frequently where strict linear paths dominate in a network such as KEGG. Chowdhury et al. (2010) <xref ref-type=\"bibr\" rid=\"pone.0084227-Chowdhury1\">[14]</xref> use the set-cover-based algorithm where either all the positive or negative samples have to be covered. If the network contains small components, then it might not be able to cover the positive or negative samples completely. Dao et al. (2010) <xref ref-type=\"bibr\" rid=\"pone.0084227-Dao1\">[15]</xref> utilize density-constrained clustering which would not work well on sparse networks such as KEGG. Su et al. (2010) <xref ref-type=\"bibr\" rid=\"pone.0084227-Su1\">[16]</xref> look for the discriminative linear paths and then combine them to build the subnetwork. In a sparse network, the linear paths may not intersect each other. In contrast, we showed that COSSY works well on networks with highly different topologies such as KEGG and STRING. The average LOOCV accuracies of COSSY are 93.2% and 92.7% when used with KEGG and STRING, respectively, and are high and close to each other.</p>\n",
       "</sec>\n",
       "<sec id=\"s4e\">\n",
       "<title>Subnetworks with Heterogeneous Characteristics</title>\n",
       "<p>Chuang et al. (2007) <xref ref-type=\"bibr\" rid=\"pone.0084227-Chuang1\">[13]</xref> take the sum of the expressions of all the nodes in a subnetwork to aggregate its expressions, and thus the process favors the subnetworks with homogeneous characteristics where all the nodes are either positively or negatively expressed. The NetCover algorithm <xref ref-type=\"bibr\" rid=\"pone.0084227-Chowdhury1\">[14]</xref> also aggregates the expressions in the same way and suffers from the same problem. Instead of aggregating the expressions, COSSY uses clustering and thus it is capable of utilizing the heterogeneous nature to find the subnetworks that have both positively and negatively expressed nodes. For example, the representative probeset of the Leukemia MIS in <xref ref-type=\"fig\" rid=\"pone-0084227-g004\">Figure 4</xref> contains one positively and four negatively expressed probes.</p>\n",
       "</sec>\n",
       "</sec>\n",
       "<sec id=\"s5\">\n",
       "<title>Conclusion</title>\n",
       "<p>In this paper, we have introduced a non-greedy algorithm, COSSY, to partition a molecular interaction network and find the important subnetworks discriminating between two phenotypes. COSSY can find subnetworks with heterogeneous characteristics in any network irrespective of the topology. Its accuracy is comparable to the state-of-the-art classifiers. We also illustrated an interpretation of the results to discover the development process of a disease.</p>\n",
       "</sec>\n",
       "<sec id=\"s6\" sec-type=\"supplementary-material\">\n",
       "<title>Supporting Information</title>\n",
       "<supplementary-material content-type=\"local-data\" id=\"pone.0084227.s001\">\n",
       "<label>Document S1</label>\n",
       "<caption>\n",
       "<p>\n",
       "<bold>Supporting information document.</bold> This document contains seven sections (S1. Estimated t-score, S2. Binary Vote, S3. Dataset Download Sources, S4. An Experiment with the appropriate range, S5. Algorithm Illustration, S6. Irregular Network Topology, and S7. Representative probesets of the top MISs), two figures (S1. An illustration of the MIS generation, and S2. Topology of out-of-size MISs), and five Tables (S1. Dataset Download Sources, S2. LOOCV accuracy with different appropriate ranges from KEGG, S3. LOOCV accuracy with different appropriate ranges from STRING, S4. Representative probesets of the top MISs from KEGG, and S5. Representative probesets of the top MISs from STRING).</p>\n",
       "<p>(PDF)</p>\n",
       "</caption>\n",
       "<media xlink:href=\"pone.0084227.s001.pdf\">\n",
       "<caption>\n",
       "<p>Click here for additional data file.</p>\n",
       "</caption>\n",
       "</media>\n",
       "</supplementary-material>\n",
       "</sec>\n",
       "<back>\n",
       "<ref-list>\n",
       "<title>References</title>\n",
       "<ref id=\"pone.0084227-Risch1\">\n",
       "<label>1</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Risch</surname><given-names>NJ</given-names></name> (<year>2000</year>) <article-title>Searching for genetic determinants in the new millennium</article-title>. <source>Nature</source>\n",
       "<volume>405</volume>: <fpage>847</fpage>–<lpage>856</lpage>.<pub-id pub-id-type=\"pmid\">10866211</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Marchini1\">\n",
       "<label>2</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Marchini</surname><given-names>J</given-names></name>, <name><surname>Donnelly</surname><given-names>P</given-names></name>, <name><surname>Cardon</surname><given-names>LR</given-names></name> (<year>2005</year>) <article-title>Genome-wide strategies for detecting multiple loci that influence complex diseases</article-title>. <source>Nature Genetics</source>\n",
       "<volume>37</volume>: <fpage>413</fpage>–<lpage>417</lpage>.<pub-id pub-id-type=\"pmid\">15793588</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Schadt1\">\n",
       "<label>3</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Schadt</surname><given-names>EE</given-names></name> (<year>2009</year>) <article-title>Molecular networks as sensors and drivers of common human diseases</article-title>. <source>Nature</source>\n",
       "<volume>461</volume>: <fpage>218</fpage>–<lpage>223</lpage>.<pub-id pub-id-type=\"pmid\">19741703</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Geman1\">\n",
       "<label>4</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Geman</surname><given-names>D</given-names></name>, <name><surname>D’Avignon</surname><given-names>C</given-names></name>, <name><surname>Naiman</surname><given-names>DQ</given-names></name>, <name><surname>Winslow</surname><given-names>RL</given-names></name> (<year>2004</year>) <article-title>Classifying gene expression profiles from pairwise mRNA comparisons</article-title>. <source>Statistical Applications in Genetics and Molecular Biology</source>\n",
       "<volume>3</volume>: <fpage>Article19</fpage>.<pub-id pub-id-type=\"pmid\">16646797</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Tan1\">\n",
       "<label>5</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Tan</surname><given-names>AC</given-names></name>, <name><surname>Naiman</surname><given-names>DQ</given-names></name>, <name><surname>Xu</surname><given-names>L</given-names></name>, <name><surname>Winslow</surname><given-names>RL</given-names></name>, <name><surname>Geman</surname><given-names>D</given-names></name> (<year>2005</year>) <article-title>Simple decision rules for classifying human cancers from gene expression profiles</article-title>. <source>Bioinformatics</source>\n",
       "<volume>21</volume>: <fpage>3896</fpage>–<lpage>3904</lpage>.<pub-id pub-id-type=\"pmid\">16105897</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Chopra1\">\n",
       "<label>6</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Chopra</surname><given-names>P</given-names></name>, <name><surname>Lee</surname><given-names>J</given-names></name>, <name><surname>Kang</surname><given-names>J</given-names></name>, <name><surname>Lee</surname><given-names>S</given-names></name> (<year>2010</year>) <article-title>Improving cancer classification accuracy using gene pairs</article-title>. <source>PLoS ONE</source>\n",
       "<volume>5</volume>: <fpage>e14305</fpage>.<pub-id pub-id-type=\"pmid\">21200431</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Lin1\">\n",
       "<label>7</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Lin</surname><given-names>X</given-names></name>, <name><surname>Afsari</surname><given-names>B</given-names></name>, <name><surname>Marchionni</surname><given-names>L</given-names></name>, <name><surname>Cope</surname><given-names>L</given-names></name>, <name><surname>Parmigiani</surname><given-names>G</given-names></name>, <etal>et al</etal> (<year>2009</year>) <article-title>The ordering of expression among a few genes can provide simple cancer biomarkers and signal BRCA1 mutations</article-title>. <source>BMC bioinformatics</source>\n",
       "<volume>10</volume>: <fpage>256</fpage>.<pub-id pub-id-type=\"pmid\">19695104</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Magis1\">\n",
       "<label>8</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Magis</surname><given-names>AT</given-names></name>, <name><surname>Price</surname><given-names>ND</given-names></name> (<year>2012</year>) <article-title>The top-scoring ‘N’ algorithm: a generalized relative expression classification method from small numbers of biomolecules</article-title>. <source>BMC bioinformatics</source>\n",
       "<volume>13</volume>: <fpage>227</fpage>.<pub-id pub-id-type=\"pmid\">22966958</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Wang1\">\n",
       "<label>9</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Wang</surname><given-names>H</given-names></name>, <name><surname>Zhang</surname><given-names>H</given-names></name>, <name><surname>Dai</surname><given-names>Z</given-names></name>, <name><surname>Chen</surname><given-names>Ms</given-names></name>, <name><surname>Yuan</surname><given-names>Z</given-names></name> (<year>2013</year>) <article-title>TSG: a new algorithm for binary and multi-class cancer classification and informative genes selection</article-title>. <source>BMC medical genomics</source>\n",
       "<volume>6</volume> Suppl 1<fpage>S3</fpage>.</mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Subramanian1\">\n",
       "<label>10</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Subramanian</surname><given-names>A</given-names></name>, <name><surname>Tamayo</surname><given-names>P</given-names></name>, <name><surname>Mootha</surname><given-names>VK</given-names></name>, <name><surname>Mukherjee</surname><given-names>S</given-names></name>, <name><surname>Ebert</surname><given-names>BL</given-names></name>, <etal>et al</etal> (<year>2005</year>) <article-title>Gene set enrichment analysis: A knowledge-based approach for interpreting genome-wide expression profiles</article-title>. <source>Proceedings of the National Academy of Sciences of the United States of America</source>\n",
       "<volume>102</volume>: <fpage>15545</fpage>–<lpage>15550</lpage>.<pub-id pub-id-type=\"pmid\">16199517</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Kanehisa1\">\n",
       "<label>11</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Kanehisa</surname><given-names>M</given-names></name>, <name><surname>Goto</surname><given-names>S</given-names></name>, <name><surname>Sato</surname><given-names>Y</given-names></name>, <name><surname>Furumichi</surname><given-names>M</given-names></name>, <name><surname>Tanabe</surname><given-names>M</given-names></name> (<year>2012</year>) <article-title>KEGG for integration and interpretation of large-scale molecular data sets</article-title>. <source>Nucleic acids research</source>\n",
       "<volume>40</volume>: <fpage>D109</fpage>–<lpage>14</lpage>.<pub-id pub-id-type=\"pmid\">22080510</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Franceschini1\">\n",
       "<label>12</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Franceschini</surname><given-names>A</given-names></name>, <name><surname>Szklarczyk</surname><given-names>D</given-names></name>, <name><surname>Frankild</surname><given-names>S</given-names></name>, <name><surname>Kuhn</surname><given-names>M</given-names></name>, <name><surname>Simonovic</surname><given-names>M</given-names></name>, <etal>et al</etal> (<year>2013</year>) <article-title>STRING v9.1: protein-protein interaction networks, with increased coverage and integration</article-title>. <source>Nucleic acids research</source>\n",
       "<volume>41</volume>: <fpage>D808</fpage>–<lpage>15</lpage>.<pub-id pub-id-type=\"pmid\">23203871</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Chuang1\">\n",
       "<label>13</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Chuang</surname><given-names>HY</given-names></name>, <name><surname>Lee</surname><given-names>E</given-names></name>, <name><surname>Liu</surname><given-names>YT</given-names></name>, <name><surname>Lee</surname><given-names>D</given-names></name>, <name><surname>Ideker</surname><given-names>T</given-names></name> (<year>2007</year>) <article-title>Network-based classification of breast cancer metastasis</article-title>. <source>Molecular Systems Biology</source>\n",
       "<volume>3</volume>: <fpage>140</fpage>.<pub-id pub-id-type=\"pmid\">17940530</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Chowdhury1\">\n",
       "<label>14</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Chowdhury</surname><given-names>SA</given-names></name>, <name><surname>Koyutürk</surname><given-names>M</given-names></name> (<year>2010</year>) <article-title>Identification of coordinately dysregulated subnetworks in complex phenotypes</article-title>. <source>Pacific Symposium On Biocomputing</source>\n",
       "<volume>144</volume>: <fpage>133</fpage>–<lpage>144</lpage>.</mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Dao1\">\n",
       "<label>15</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Dao</surname><given-names>P</given-names></name>, <name><surname>Colak</surname><given-names>R</given-names></name>, <name><surname>Salari</surname><given-names>R</given-names></name>, <name><surname>Moser</surname><given-names>F</given-names></name>, <name><surname>Davicioni</surname><given-names>E</given-names></name>, <etal>et al</etal> (<year>2010</year>) <article-title>Inferring cancer subnetwork markers using density-constrained biclustering</article-title>. <source>Bioinformatics</source>\n",
       "<volume>26</volume>: <fpage>i625</fpage>–<lpage>i631</lpage>.<pub-id pub-id-type=\"pmid\">20823331</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Su1\">\n",
       "<label>16</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Su</surname><given-names>J</given-names></name>, <name><surname>Yoon</surname><given-names>BJ</given-names></name>, <name><surname>Dougherty</surname><given-names>ER</given-names></name> (<year>2010</year>) <article-title>Identification of diagnostic subnetwork markers for cancer in human protein-protein interaction network</article-title>. <source>BMC Bioinformatics</source>\n",
       "<volume>11</volume>: <fpage>S8</fpage>.</mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Eddy1\">\n",
       "<label>17</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Eddy</surname><given-names>JA</given-names></name>, <name><surname>Hood</surname><given-names>L</given-names></name>, <name><surname>Price</surname><given-names>ND</given-names></name>, <name><surname>Geman</surname><given-names>D</given-names></name> (<year>2010</year>) <article-title>Identifying Tightly Regulated and Variably Expressed Networks by Differential Rank Conservation (DIRAC)</article-title>. <source>PLoS Computational Biology</source>\n",
       "<volume>6</volume>: <fpage>17</fpage>.</mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Clauset1\">\n",
       "<label>18</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Clauset</surname><given-names>A</given-names></name>, <name><surname>Newman</surname><given-names>MEJ</given-names></name>, <name><surname>Moore</surname><given-names>C</given-names></name> (<year>2004</year>) <article-title>Finding community structure in very large networks</article-title>. <source>Physical Review E</source>\n",
       "<volume>70</volume>: <fpage>1</fpage>–<lpage>6</lpage>.</mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Stalteri1\">\n",
       "<label>19</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Stalteri</surname><given-names>MA</given-names></name>, <name><surname>Harrison</surname><given-names>AP</given-names></name> (<year>2007</year>) <article-title>Interpretation of multiple probe sets mapping to the same gene in Affymetrix GeneChips</article-title>. <source>BMC Bioinformatics</source>\n",
       "<volume>8</volume>: <fpage>13</fpage>.<pub-id pub-id-type=\"pmid\">17224057</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Golub1\">\n",
       "<label>20</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Golub</surname><given-names>TR</given-names></name>, <name><surname>Slonim</surname><given-names>DK</given-names></name>, <name><surname>Tamayo</surname><given-names>P</given-names></name>, <name><surname>Huard</surname><given-names>C</given-names></name>, <name><surname>Gaasenbeek</surname><given-names>M</given-names></name>, <etal>et al</etal> (<year>1999</year>) <article-title>Molecular classification of cancer: class discovery and class prediction by gene expression monitoring</article-title>. <source>Science</source>\n",
       "<volume>286</volume>: <fpage>531</fpage>–<lpage>537</lpage>.<pub-id pub-id-type=\"pmid\">10521349</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Pomeroy1\">\n",
       "<label>21</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Pomeroy</surname><given-names>SL</given-names></name>, <name><surname>Tamayo</surname><given-names>P</given-names></name>, <name><surname>Gaasenbeek</surname><given-names>M</given-names></name>, <name><surname>Sturla</surname><given-names>LM</given-names></name>, <name><surname>Angelo</surname><given-names>M</given-names></name>, <etal>et al</etal> (<year>2002</year>) <article-title>Prediction of central nervous system embryonal tumour outcome based on gene expression</article-title>. <source>Nature</source>\n",
       "<volume>415</volume>: <fpage>436</fpage>–<lpage>442</lpage>.<pub-id pub-id-type=\"pmid\">11807556</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Shipp1\">\n",
       "<label>22</label>\n",
       "<mixed-citation publication-type=\"other\">Shipp MA, Ross KN, Tamayo P, Weng AP, Kutok JL, <etal>et al</etal>.. (2002) Diffuse large B-cell lymphoma outcome prediction by gene-expression profiling and supervised machine learning. Technical Report 1, Dana-Farber Cancer Institute, Harvard Medical School, Boston, Massachusetts, USA. margaret shipp@dfci.harvard.edu.</mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Singh1\">\n",
       "<label>23</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Singh</surname><given-names>D</given-names></name>, <name><surname>Febbo</surname><given-names>PG</given-names></name>, <name><surname>Ross</surname><given-names>K</given-names></name>, <name><surname>Jackson</surname><given-names>DG</given-names></name>, <name><surname>Manola</surname><given-names>J</given-names></name>, <etal>et al</etal> (<year>2002</year>) <article-title>Gene expression correlates of clinical prostate cancer behavior</article-title>. <source>Cancer Cell</source>\n",
       "<volume>1</volume>: <fpage>203</fpage>–<lpage>209</lpage>.<pub-id pub-id-type=\"pmid\">12086878</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Welsh1\">\n",
       "<label>24</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Welsh</surname><given-names>JB</given-names></name>, <name><surname>Sapinoso</surname><given-names>LM</given-names></name>, <name><surname>Su</surname><given-names>AI</given-names></name>, <name><surname>Kern</surname><given-names>SG</given-names></name>, <name><surname>Wang-rodriguez</surname><given-names>J</given-names></name>, <etal>et al</etal> (<year>2001</year>) <article-title>Analysis of Gene Expression Identifies Candidate Markers and Pharmacological Targets in Prostate Cancer</article-title>. <source>Cancer Research</source>\n",
       "<volume>61</volume>: <fpage>5974</fpage>–<lpage>5978</lpage>.<pub-id pub-id-type=\"pmid\">11507037</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Gordon1\">\n",
       "<label>25</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Gordon</surname><given-names>GJ</given-names></name>, <name><surname>Jensen</surname><given-names>RV</given-names></name>, <name><surname>Hsiao</surname><given-names>LL</given-names></name>, <name><surname>Gullans</surname><given-names>SR</given-names></name>, <name><surname>Blumenstock</surname><given-names>JE</given-names></name>, <etal>et al</etal> (<year>2002</year>) <article-title>Translation of microarray data into clinically relevant cancer diagnostic tests using gene expression ratios in lung cancer and mesothelioma</article-title>. <source>Cancer Research</source>\n",
       "<volume>62</volume>: <fpage>4963</fpage>–<lpage>4967</lpage>.<pub-id pub-id-type=\"pmid\">12208747</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Ramaswamy1\">\n",
       "<label>26</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Ramaswamy</surname><given-names>S</given-names></name>, <name><surname>Tamayo</surname><given-names>P</given-names></name>, <name><surname>Rifkin</surname><given-names>R</given-names></name>, <name><surname>Mukherjee</surname><given-names>S</given-names></name>, <name><surname>Yeang</surname><given-names>CH</given-names></name>, <etal>et al</etal> (<year>2001</year>) <article-title>Multiclass cancer diagnosis using tumor gene expression signatures</article-title>. <source>Proceedings of the National Academy of Sciences of the United States of America</source>\n",
       "<volume>98</volume>: <fpage>15149</fpage>–<lpage>15154</lpage>.<pub-id pub-id-type=\"pmid\">11742071</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Ward1\">\n",
       "<label>27</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Ward</surname><given-names>JH</given-names></name> (<year>1963</year>) <article-title>Hierarchical grouping to optimize an objective function</article-title>. <source>Journal of the American Statistical Association</source>\n",
       "<volume>58</volume>: <fpage>236</fpage>–<lpage>244</lpage>.</mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Mardia1\">\n",
       "<label>28</label>\n",
       "<mixed-citation publication-type=\"other\">Mardia KV, Kent JT, Bibby JM (1979) Multivariate analysis. Academic Press.</mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Mudunuri1\">\n",
       "<label>29</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Mudunuri</surname><given-names>U</given-names></name>, <name><surname>Che</surname><given-names>A</given-names></name>, <name><surname>Yi</surname><given-names>M</given-names></name>, <name><surname>Stephens</surname><given-names>RM</given-names></name> (<year>2009</year>) <article-title>bioDBnet: the biological database network</article-title>. <source>Bioinformatics</source>\n",
       "<volume>25</volume>: <fpage>555</fpage>–<lpage>556</lpage>.<pub-id pub-id-type=\"pmid\">19129209</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-DosSantos1\">\n",
       "<label>30</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Dos Santos</surname><given-names>C</given-names></name>, <name><surname>Demur</surname><given-names>C</given-names></name>, <name><surname>Bardet</surname><given-names>V</given-names></name>, <name><surname>Prade-Houdellier</surname><given-names>N</given-names></name>, <name><surname>Payrastre</surname><given-names>B</given-names></name>, <etal>et al</etal> (<year>2008</year>) <article-title>A critical role for Lyn in acute myeloid leukemia</article-title>. <source>Blood</source>\n",
       "<volume>111</volume>: <fpage>2269</fpage>–<lpage>79</lpage>.<pub-id pub-id-type=\"pmid\">18056483</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Park1\">\n",
       "<label>31</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Park</surname><given-names>S</given-names></name>, <name><surname>Chapuis</surname><given-names>N</given-names></name>, <name><surname>Tamburini</surname><given-names>J</given-names></name>, <name><surname>Bardet</surname><given-names>V</given-names></name>, <name><surname>Cornillet-Lefebvre</surname><given-names>P</given-names></name>, <etal>et al</etal> (<year>2010</year>) <article-title>Role of the PI3K/AKT and mTOR signaling pathways in acute myeloid leukemia</article-title>. <source>Haematologica</source>\n",
       "<volume>95</volume>: <fpage>819</fpage>–<lpage>28</lpage>.<pub-id pub-id-type=\"pmid\">19951971</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Tamburini1\">\n",
       "<label>32</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Tamburini</surname><given-names>J</given-names></name>, <name><surname>Chapuis</surname><given-names>N</given-names></name>, <name><surname>Bardet</surname><given-names>V</given-names></name>, <name><surname>Park</surname><given-names>S</given-names></name>, <name><surname>Sujobert</surname><given-names>P</given-names></name>, <etal>et al</etal> (<year>2008</year>) <article-title>Mammalian target of rapamycin (mTOR) inhibition activates phosphatidylinositol 3-kinase/Akt by up-regulating insulinlike growth factor-1 receptor signaling in acute myeloid leukemia: rationale for therapeutic inhibition of both pathways</article-title>. <source>Blood</source>\n",
       "<volume>111</volume>: <fpage>379</fpage>–<lpage>82</lpage>.<pub-id pub-id-type=\"pmid\">17878402</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Oellerich1\">\n",
       "<label>33</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Oellerich</surname><given-names>T</given-names></name>, <name><surname>Oellerich</surname><given-names>MF</given-names></name>, <name><surname>Engelke</surname><given-names>M</given-names></name>, <name><surname>Münch</surname><given-names>S</given-names></name>, <name><surname>Mohr</surname><given-names>S</given-names></name>, <etal>et al</etal> (<year>2013</year>) <article-title>\n",
       "<italic>β</italic>2 integrin-derived signals induce cell survival and proliferation of AML blasts by activating a Syk/STAT signaling axis</article-title>. <source>Blood</source>\n",
       "<volume>121</volume>: <fpage>3889</fpage>–<lpage>99</lpage>.<pub-id pub-id-type=\"pmid\">23509157</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Sukhai1\">\n",
       "<label>34</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Sukhai</surname><given-names>MA</given-names></name>, <name><surname>Prabha</surname><given-names>S</given-names></name>, <name><surname>Hurren</surname><given-names>R</given-names></name>, <name><surname>Rutledge</surname><given-names>AC</given-names></name>, <name><surname>Lee</surname><given-names>AY</given-names></name>, <etal>et al</etal> (<year>2013</year>) <article-title>Lysosomal disruption preferentially targets acute myeloid leukemia cells and progenitors</article-title>. <source>The Journal of clinical investigation</source>\n",
       "<volume>123</volume>: <fpage>315</fpage>–<lpage>28</lpage>.<pub-id pub-id-type=\"pmid\">23202731</pub-id></mixed-citation>\n",
       "</ref>\n",
       "<ref id=\"pone.0084227-Sing1\">\n",
       "<label>35</label>\n",
       "<mixed-citation publication-type=\"journal\">\n",
       "<name><surname>Sing</surname><given-names>T</given-names></name>, <name><surname>Sander</surname><given-names>O</given-names></name>, <name><surname>Beerenwinkel</surname><given-names>N</given-names></name>, <name><surname>Lengauer</surname><given-names>T</given-names></name> (<year>2005</year>) <article-title>ROCR: visualizing classifier performance in R. Bioinformatics (Oxford, England)</article-title>. <volume>21</volume>: <fpage>3940</fpage>–<lpage>1</lpage>.</mixed-citation>\n",
       "</ref>\n",
       "</ref-list>\n",
       "</back>\n",
       "</article>\n",
       "</pmc-articleset></body></html>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "page_soup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Get full text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\n\\n\\n\\nPLoS One\\nPLoS ONE\\nplos\\nplosone\\n\\nPLoS ONE\\n\\n1932-6203\\n\\nPublic Library of Science\\nSan Francisco, USA\\n\\n\\n\\n24392115\\n3877685\\nPONE-D-13-37281\\n10.1371/journal.pone.0084227\\n\\n\\nResearch Article\\n\\n\\nBiology\\n\\nComputational Biology\\n\\nMolecular Genetics\\n\\nGene Classes\\nGene Identification and Analysis\\nGene Expression\\n\\n\\n\\nMetabolic Networks\\nMicroarrays\\nRegulatory Networks\\nSignaling Networks\\n\\n\\n\\nGenetics\\n\\nHuman Genetics\\n\\nGenetic Association Studies\\n\\n\\n\\nGene Networks\\nGenetics of Disease\\n\\n\\n\\n\\nComputer Science\\n\\nAlgorithms\\n\\n\\n\\n\\nAutomatic Context-Specific Subnetwork Discovery from Large Interaction Networks\\nContext-Specific Subnetwork Discovery\\n\\n\\n\\n\\nSaha\\nAshis\\n\\n\\n1\\n\\n\\n\\n\\nTan\\nAik Choon\\n\\n\\n2\\n\\n\\n\\n\\nKang\\nJaewoo\\n\\n\\n1\\n\\n\\n3\\n\\n\\n*\\n\\n\\n\\n\\n1\\nDepartment of Computer Science and Engineering, Korea University, Seoul, Korea\\n\\n\\n2\\nDepartment of Medicine/Medical Oncology, University of Colorado Anschutz Medical Campus, Aurora, Colorado, United States of America\\n\\n\\n3\\nInterdisciplinary Graduate Program in Bioinformatics, Korea University, Seoul, Korea\\n\\n\\n\\n\\nAloy\\nPatrick\\n\\nEditor\\n\\n\\n\\n\\nInstitute for Research in Biomedicine, Spain\\n\\n\\n* E-mail: kangj@korea.ac.kr\\n\\nCompeting Interests: The authors have declared that no competing interests exist.\\n\\n\\nConceived and designed the experiments: AS JK. Performed the experiments: AS. Analyzed the data: AS ACT JK. Wrote the paper: AS ACT JK.\\n\\n\\n\\n2014\\n\\n\\n1\\n1\\n2014\\n\\n9\\n1\\ne84227\\n\\n\\n10\\n9\\n2013\\n\\n\\n21\\n11\\n2013\\n\\n\\n\\n© 2014 Saha et al\\n2014\\nSaha et al\\n\\nThis is an open-access article distributed under the terms of the Creative Commons Attribution License, which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are properly credited.\\n\\n\\n\\nGenes act in concert via specific networks to drive various biological processes, including progression of diseases such as cancer. Under different phenotypes, different subsets of the gene members of a network participate in a biological process. Single gene analyses are less effective in identifying such core gene members (subnetworks) within a gene set/network, as compared to gene set/network-based analyses. Hence, it is useful to identify a discriminative classifier by focusing on the subnetworks that correspond to different phenotypes. Here we present a novel algorithm to automatically discover the important subnetworks of closely interacting molecules to differentiate between two phenotypes (context) using gene expression profiles. We name it COSSY (COntext-Specific Subnetwork discoverY). It is a non-greedy algorithm and thus unlikely to have local optima problems. COSSY works for any interaction network regardless of the network topology. One added benefit of COSSY is that it can also be used as a highly accurate classification platform which can produce a set of interpretable features.\\n\\n\\nThis work was supported by the National Research Foundation of Korea (NRF) grant funded by the Korean government (MSIP) (2012R1A2A2A01014729); and Next-Generation Information Computing Development Program through the National Research Foundation of Korea (NRF) funded by the Ministry of Science, ICT & Future Planning (2012M3C4A7033341). The funders had no role in study design, data collection and analysis, decision to publish, or preparation of the manuscript.\\n\\n\\n\\n\\n\\n\\n\\nIntroduction\\nBiological systems are complex in nature; individual molecular components (e.g., genes and proteins) interact with each other in specific networks to exert their functions. The development of new technologies and powerful computational algorithms to sequence and characterize genomes have enabled researchers to acquire and analyze tens of thousands of ‘omic’ data points across the genetic and epigenetic changes within biological systems. One of the main difficulties in the gene expression analysis is the size of the ‘omic’ data. Due to the large number of genes, the complexity of the algorithms becomes very high.\\nThe common practice is to rank the individual genes according to a similarity or dissimilarity score and then use the important genes for prediction. However, Risch (2000) suggested that a combination of genes is more apt for common disease classification than individual genes [1]. Marchini, Donnelly, and Cardon (2005) illustrated the effect that gene-gene interactions have on complex diseases [2]. Moreover, recent research suggests that molecular networks, rather than random sets of genes, can help find underlying physiological states associated with disease and understand complex biological processes [3]. Such approaches are more objective and robust in their ability to discover sets of coordinated differentially expressed genes among pathway members and their association to a specific biological phenotype. These analyses may provide new insights linking biological phenotypes to their underlying molecular mechanisms, as well as suggesting new hypotheses about the pathway membership and connectivity.\\nFor this reason, researchers are now concentrating on gene-gene interactions instead of on individual gene analyses. Geman et al. (2004) introduced the relative expression reversal concept of top-scoring pair (TSP) as a classification approach for gene expression profiles [4]. They attempted to figure out the pairs of genes whose expression levels typically invert from one class to another. This concept has been improved by majority voting of the  top scoring pairs (k-TSP) [5]. Chopra et al. (2010) also introduced their version of gene-pairing which they call doublet\\n[6]. Instead of using only the inverse relations of the expression levels, they provided a framework for using different kinds of relations between two genes. Methods that extended the same concept for finding triplets [7] and other combinations of gene pairs [8], [9] have been developed.\\nAlthough these methods produce highly accurate results, they only consider the interaction of a few genes at a time which is still a small number to imply the biological process of a disease. Conversely, gene set enrichment analysis methods such as the Gene Set Enrichment Analysis (GSEA) [10] have improved the ability to identify candidate genes that are correlated with a disease state by exploiting the idea that gene expression alterations might be revealed at the level of biological pathways or co-regulated gene sets, rather than at the level of individual genes. However, the interactions (or topology) of the identified subset of genes within the predefined gene sets are not captured by these methods.\\nThe availability of large-scale molecular interaction data such as those collected in KEGG (Kyoto Encyclopedia of Genes and Genomes) [11] and STRING (Search Tool for the Retrieval of Interacting Genes/Proteins) [12] provides an opportunity to incorporate these interactions into omic data analysis. The advantage of these network-based approaches is that they can fully describe the complex biological processes driving disease phenotypes. Most network-based approaches attempted to identify informative subnetworks - a set of genes that interact directly and have differential expression patterns - for disease classification. These approaches attempted to find these subnetworks using a greedy approach, starting with a small network (typically a single node) and adding nodes greedily until the score is improved [13]–[16]. However, they may suffer from local-optima problems and may not even work for a sparse network such as KEGG because their greedy expansion process could halt imgmaturely due to sparse connectivity. Alternatively, some approaches, such as DIRAC (Differential Rank Conservation)[17], utilized user-defined and pre-partitioned subnetworks for classification.\\nHere we propose a novel COntext-Specific Subnetwork discoverY (COSSY) algorithm to automatically discover the important subnetworks to differentiate between two phenotypes (context). COSSY finds the important subnetworks using a non-greedy approach. Unlike the greedy approach, our approach first divides the network, and then ranks the subnetworks based on the biological context of interest. We demonstrate that COSSY works for any network regardless of its topology. We show that COSSY is capable of identifying discriminative subnetworks while achieving classification results comparable to those of other state-of-the-art methods. Finally, we illustrate that the subnetworks identified by COSSY provide meaningful biological insights from high-throughput datasets.\\n\\nCOSSY Algorithm\\nLet  be a molecular interaction network (graph) where each vertex in  denotes a molecule (or a set of molecules) and each edge in  denotes an interaction between molecules. A molecule can be a gene, a protein, a nucleic acid or a similar biochemical object. Let  be the training dataset containing the microarray expression profiles of  samples, where each profile consists of  probes.  can be represented as an  matrix where  denotes the expression value of the -th probe, , in the -th sample, . The row vector  thus represents the -th probe across  samples and the column vector  represents the -th sample across  probes. Let  be the vector of the class labels (context) for  samples where . Our task is to find the subnetworks (subgraphs) of  to predict the class (context) of a new sample .\\nTo give an overview, we first partition the molecular interaction network  to generate the molecular interaction subnetworks. Each subnetwork is then ranked using a clustering approach and finally the top  subnetworks vote to predict the context of the new sample,  (Figure 1).\\n\\n10.1371/journal.pone.0084227.g001\\nFigure 1\\n\\nOverview of COSSY.\\nCommunities are extracted from the molecular interaction network to generate Molecular Interaction Subnetworks (MISs)(A–B). Each MIS is mapped to microarray probes and all the samples are clustered according to the expression pattern of a certain number of highly differentially expressed probes (3 probes in this example figure). MISs are ranked by the entropy score which is lowest when every cluster contains only one type (phenotype) of samples (C–D). Finally, the top  MISs cast votes to predict the context (phenotype) of a new sample. The voting depends on the proportion of different types of samples in the cluster closest to the new sample (E–G).\\n\\n\\n\\n\\n\\nMolecular Interaction Subnetwork (MIS)\\nWe use a biological knowledge-base in the form of a network in order to get biologically interpretable results. An important property of such a network is that it shows how a group of molecules interact with each other, often performing a molecular function. However, it is not necessary for all the molecules in a network to interact with each other. For example, some of the genes in the thyroid cancer pathway in KEGG (id: hsa05216) are not connected (Figure 2). Traditional pathway analyses consider all the molecules (genes) in the same pathway together. We believe that the non-interacting set of molecules should be considered separately. Therefore, we propose to include only the interacting molecules in a candidate set, which we call the Molecular Interaction Subnetwork (MIS).\\n\\n10.1371/journal.pone.0084227.g002\\nFigure 2\\n\\nThe thyroid cancer pathway in KEGG (ID: hsa05216).\\nThe top rectangle marked by C1 shows that REDPTC and TRK have an indirect effect on the activation of Ras. Ras then activates BRAF, and BRAF phosphorylates MEK which in turn phosphorylates ERK. The result of this path is proliferation survival. This pathway has five connected components (C1–C5). Among them, C2 is actually a subset of C1, and the others are fully disconnected, i.e., there is no significant interaction between any pair of the components.\\n\\n\\n\\nWe further adjust the idea of an MIS for large number vertices in a connected network. Some pathways such as the cell cycle pathway in KEGG (id: hsa04110) contain a large number of genes all of which are connected to each other, although different parts of the pathway are annotated to have different functions. We believe that, it would not be wise to assume that each and every molecule in a large network responsible for a process would also be responsible for another process. Instead, we reasonably assume that a subset of molecules in one process may also affect the development of another process. Hence, we propose to divide the large network into pieces so that only the closely interacting molecules are in the same subnetwork. An MIS will contain all the molecules in such a subnetwork.\\nWe utilize the community structure of a network –groups of nodes within which the edge densities are high, but between which they are low – to find the closely interacting subnetworks. We first build a community dendrogram as described by [4]. Smaller communities can be made by splitting the dendrogram from the top. Modularity, a measure of the quality of a particular division of a network, is based on the ratio of the number of edges within groups to the total number of edges in the network. The division improves as the modularity increases. Among all the available ones, the pair of groups of nodes whose integration would produce the highest modularity are joined until one group remains. This process can be presented as a dendrogram whose division from the top to the bottom would create smaller communities. If the community size is within an appropriate range [minRange, maxRange], we consider it as an appropriate community. If the size is above the appropriate range, we cut the community dendrogram from the top. So we get two small communities and we recursively do this until the size is within the appropriate range. The partition process discards any branch whose size goes below the appropriate range and once the partition process is over, each node in the discarded branch is individually merged with the closest appropriate community. This merging step may cause some communities to grow above the appropriate range. For each of these communities, we create a new network taking all the nodes of the community and the edges among them from the original network, and then we apply the whole process to the new network. Algorithm 1, along with Algorithm 2–4, describes the MIS generation process in detail. The algorithm is illustrated in Figure S1 in Supporting Document S1.\\n\\n\\nAlgorithm 1 Generate MIS from network\\n\\nInput:\\n, a molecular interaction network\\n\\nOutput: List of MISs generated from \\n\\n1: for all Connected component  in \\ndo\\n\\n2: if\\n\\nthen\\n\\n3: Generate MISs from  using Algorithm 2 and save them.\\n4: eles\\n\\n5: Generate an MIS taking all the molecules in  and save it\\n6: end if\\n\\n7: end for\\n\\n8: return all saved MISs\\n\\n\\n\\n\\nAlgorithm 2 Generate MIS from connected network\\n\\nInput:\\nInput:\\n, a connected network with at least  nodes.\\n\\nOutput: MISs generated from .\\n1:  Build a community dendrogram as described by [18].\\n2:  appropriate communities generated from  using Algorithm 3.\\n3:  communities after merging the discarded nodes in step 2 with the closest appropriate communities using Algorithm 4.\\n4: for all community  in \\ndo.\\n\\n5: if size()\\n\\nthen.\\n\\n6: Create and save an MIS taking all the molecules in .\\n7: else.\\n\\n8: \\n a subgraph of  containing all the leaf nodes of  and all the edges among them in .\\n9: Recursively generate and save MISs from .\\n10: end if.\\n\\n11: end for.\\n\\n12: return all saved MISs.\\n\\n\\n\\n\\nAlgorithm 3 Extract appropriate communities from dendrogram\\n\\nInput:\\n, the community dendrogram.\\n\\nOutput: Appropriate communities extracted from .\\n1: \\n No. of leaf nodes in .\\n2: if\\n\\nthen.\\n\\n3: Create a community of all the leaf nodes in  and save it.\\n4: even if\\n\\nthen.\\n\\n5: Divide  into 2 parts from the top -  and .\\n6: Recursively extract and save appropriate communities from  and .\\n7: eles.\\n\\n8: Discard the .\\n9: end if.\\n\\n10: return saved appropriate communities.\\n\\n\\n\\n\\nAlgorithm 4 Merge discarded nodes with given communities\\n\\nInput:\\n, nodes not taken in any community; , given communities.\\n\\nOutput: Communities with  merged.\\n1: repeat.\\n\\n2:  The subset of  directly connected (1-hop distance) to some node in  from the original network.\\n3: for all node  in \\ndo.\\n\\n4: .\\n5: .\\n6: loop.\\n\\n7:  Count no. of -hop distant nodes (from ) in each community in .  communities having \\n-hop distant nodes from .\\n9: if\\n \\u200a=\\u200a0 then.\\n\\n10: Mark 's membership to one community in  randomly.\\n11: break loop.\\n12: else if\\n \\u200a=\\u200a1 then.\\n\\n13: Mark 's membership to the only community in .\\n14: break loop.\\n15: else.\\n\\n16: .\\n17: .\\n18: end if.\\n\\n19: end loop.\\n\\n20: end for.\\n\\n21:  communities after merging every node in  according to its membership.\\n22: .\\n23: until\\n.\\n24: return\\n.\\n\\n\\n\\n\\nFeature Selection\\nIt is necessary to map the molecules to the probes used in . If one molecule is mapped to multiple probes we include all of them separately following the suggestion to consider each probe independently [19]. An MIS is selected if its molecules are mapped to at least a certain number of unique probes. Thus we avoid the unreliable effect of a single or low number of probes. For every selected MIS, we propose to take the same number of probes to consider them equally in the next step. In this study, we choose the five most differentially expressed probes in each MIS which we call the representative probeset for the MIS. We estimate the differential expression of a probe using a modified version of Welch’s t-statistic score (Section S1 in Document S1).\\nAs one molecule can be mapped to multiple probes and can be in multiple nodes of some molecular interaction networks, the representative probesets of MISs may overlap with each other. We measure the overlap as the ratio of the number of common probes in the representative probesets of the two MISs over the number of probes in each representative probeset, , where RP(.) denotes the representative probeset of an MIS. We take two MISs with the highest overlap and combine them into one MIS by taking a union of all the molecules in both MISs. We continue to combine MISs with the highest overlaps as long as they are significant ( in this study).\\n\\n\\nRanking\\nWe rank the MISs based on the expression pattern of their representative probesets. We cluster all the samples using the expression values of the representative probeset. Each cluster represents a separate pattern. If a specific pattern corresponds to a class, we can expect that almost all the samples in the corresponding cluster will be of the same class (Figure 3). Based on this behavior, we propose an entropy-based score to rank the MISs.\\n\\n10.1371/journal.pone.0084227.g003\\nFigure 3\\n\\nRanking of MIS.\\nLet three probes (i, j, and k) constitute the representative probeset of an MIS. We plot all the samples with the expression values of these probes in separate dimensions, and then we cluster the samples. If the samples in a cluster are mostly of one kind (such as  or ), we can say that the cluster’s expression pattern represents the corresponding class (positive or negative). The ranking of an MIS producing such clusters should be high.\\n\\n\\n\\nLet us consider that all the samples in dataset  have been divided into  clusters  using the expression values of all the probes in an MIS representative probeset. Each cluster  contains  positive samples and  negative samples. We normalize the number of positive and negative samples in each cluster by dividing it by the total number of positive and negative samples in the dataset, respectively. The normalized number of positive samples is  and the normalized number of negative samples is . The entropy of cluster  is then defined as(1)\\n\\nThe entropy of the MIS is then defined as(2)\\n\\nThe range of each cluster’s entropy is . Consequently, the entropy of an MIS is always within . If all the samples in every cluster are of one kind (same class), then the entropy becomes zero. In contrast, it becomes one if every cluster contains an equal proportion of positive and negative samples. MISs are sorted by their entropies in an increasing order, i.e., the MISs with the lowest and highest entropies are at the top and bottom of the ranking, respectively.\\n\\n\\nContext Prediction\\nThe context (or class) of a new sample  is predicted from the weighted voting by the top  MISs .  votes for the positive and negative class with the weight of  and , respectively (Eqs. 3 and 4).(3)\\n(4)where  is the closest cluster to  among the clusters produced by , and  and  are the normalized numbers of the positive and negative samples, respectively, in . The sums of the voting weights by the top  MISs for the positive class and the negative class are  and , respectively (Eqs. 5 and 6).\\n\\n(5)\\n(6)\\n\\nThe class, with a voting weight that is higher than the other, is predicted to be the class of the new sample, . However, if both weights,  and , become equal, the class is predicted from binary voting: a class gets the full weight (i.e., one) if its number of samples surpasses that of the other class in a nearby cluster; otherwise it gets zero (Section S2 in Supporting Document S1). Binary voting would take place infrequently as we use only the top ranked MISs. In fact, it did not occur in any of our experiments. The binary voting system guarantees the winner when  is odd. The classification rule is finally expressed as below (Eq. 7).(7)\\n\\n\\n\\n\\nDatasets, Implementation and Evaluation Method\\n\\nMicroarray Datasets\\nWe tested COSSY on seven publicly available cancer datasets, as shown in Table 1. In this paper, we refer to the datasets by the names mentioned in the first column of Table 1. The purpose of the Leukemia microarray dataset is to identify genes that distinguish Acute Myeloid Leukemia (AML), a deathly disease, from Acute Lymphoblastic Leukemia (ALL) [20]. It contains expression levels of 7129 probes from 72 samples (ALL:47, AML:25). We used dataset B from [21], which contains 25 classic and 9 demoplastic medulloblastoma samples, as the CNS dataset. The samples of the DLBCL dataset are divided into 2 categories: Diffuse Large B-Cell Lymphoma (DLBCL) and Follicular Lymphoma (FL) [22]. The Prostate1 dataset, distinguishing between prostate tumors and non-tumor prostates (normal), used 102 high-quality expression profiles from [23]. The Prostate3 dataset includes 24 prostate tumor and 9 normal samples from [24]. The goal of the Lung dataset is to differentiate between Malignant Pleural Mesothelioma (MPM) and Adenocarcinoma (ADCA) of the lung [25]. Finally, the GCM_total dataset from [26], which contains 90 normal and 190 tumor samples of several types of cancer, constitutes the GCM dataset. The sources of these datasets have been listed in Table S1 in Supporting Document S1.\\n\\n10.1371/journal.pone.0084227.t001\\nTable 1\\n\\nMicroarray Datasets.\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\nDataset Name\\n#Probes\\nPositive Class (#samples)\\nNegative Class (#samples)\\nReference\\n\\n\\n\\n\\nLeukemia\\n7129\\nAML (25)\\nALL (47)\\n\\n[20]\\n\\n\\n\\nCNS\\n7129\\nDemoplastic (9)\\nClassic (25)\\n\\n[21]\\n\\n\\n\\nDLBCL\\n7129\\nDLBCL (58)\\nFL (19)\\n\\n[22]\\n\\n\\n\\nProstate1\\n12600\\nTumor (52)\\nNormal (50)\\n\\n[23]\\n\\n\\n\\nProstate3\\n12626\\nTumor (24)\\nNormal (9)\\n\\n[24]\\n\\n\\n\\nLung\\n12533\\nMPM (31)\\nADCA (150)\\n\\n[25]\\n\\n\\n\\nGCM\\n16063\\nTumor (190)\\nNormal (90)\\n\\n[26]\\n\\n\\n\\n\\n\\n\\n\\n\\nThe first column, ‘Dataset Name’, indicates the name of the microarray dataset used in the manuscript. ‘#Probes’ shows the number of probes present in the dataset. The third and fourth columns contain the name of the positive and negative class, respectively, followed by the number of samples of that class. The last column shows the reference the dataset was collected from.\\n\\n\\n\\n\\n\\nImplementation of COSSY\\nCOSSY is implemented as an R package. It requires R (2.15.3) along with utils, stats, limma, and igraph (0.6.5) packages. COSSY has been tested on Windows, Linux, and Mac operating systems. The package and data used in this study are available for download on our website: http://infos.korea.ac.kr/cossy/.\\n\\n\\nExperimental Settings\\nWe first performed the quantile normalization over the training dataset and then standardized the probe data using z-score, , where  is the expression value of the -th probe in the -th sample;  and  are the mean and standard deviation of the expression values, respectively, of the -th probe. Ward’s minimum variance agglomerative hierarchical clustering [27] was used with euclidean distance. The dendrogram produced by Ward’s method was cut into  groups to form  clusters. The number of clusters, , was chosen from the rule of thumb, , where  is the number of samples [28]. To avoid the effect of noisy data, the median of the data points in each cluster was taken as its centroid. A new data point would fall in the cluster whose centroid is the closest to the new point. Finally, , the number of the top MISs to vote, was selected experimentally. We experimented with  and used the minimum  that produced the highest accuracy.\\n\\n\\nCross Validation\\nWe validate the classification of COSSY by standard Leave-One-Out Cross-Validation (LOOCV). For each sample  in , we train the classifier using the remaining () samples and then the classifier predicts the class of . The LOOCV accuracy is the fraction of all the samples that are correctly classified.\\n\\n\\n\\nResults\\n\\nNetworks Generated from KEGG and STRING\\nWe applied COSSY on two well-known molecular interaction networks: KEGG and STRING. We chose KEGG and STRING, as they are both very different in nature. KEGG is composed of a number of disjoint pathways where the same molecule can be present in multiple pathways. In contrast, a molecule in STRING can have only one node. The nodes in KEGG may represent different types of objects (gene, gene product, chemical compound, etc.) while the nodes in STRING represent proteins. A KEGG node may have multiple molecules, while a STRING node always represents only one protein. The topologies of these two networks are also significantly different. STRING shows a much higher clustering tendency than KEGG. In this study, we used the unweighted and undirected graph of the protein networks – the network of the genes and their interactions – from the human signaling pathways in KEGG (release 62.0), and the human protein-protein interactions – both experimental and predicted with high confidence (score>0.7) – in STRING (release 9.05). A few properties of the networks are shown in Table 2.\\n\\n10.1371/journal.pone.0084227.t002\\nTable 2\\n\\nNetwork Properties of KEGG and STRING.\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\nNetwork\\nTotalNodes\\nGene (Protein) Nodes\\nTotal Edges\\nConnectedComp.\\nAvg. Node Degree\\nMax Node Degree\\nClusteringCoefficient\\n\\n\\n\\n\\nKEGG\\n19568\\n10691\\n10728\\n4494\\n1.84\\n43\\n0.19\\n\\n\\nSTRING\\n14250\\n14250\\n215800\\n182\\n30.30\\n1110\\n0.61\\n\\n\\n\\n\\n\\n\\n\\nThe ‘Total Nodes’ column contains the total number of nodes available in the network while the ‘Gene (Protein) Nodes’ column shows the number of nodes with at least one gene in KEGG (or one protein in STIRING). The fourth and fifth columns contain the total number of edges, and the number of connected components having at least one gene (or protein), respectively. ‘Avg. Node Degree’ represents the number of edges a node has on average. ‘Max Node Degree’ denotes the maximum number of edges a node has in the network. ‘Clustering Coefficient’ is the ratio of the triangles to the connected triples in a graph.\\n\\n\\n\\nAs described before, we try to keep the size of the MISs within the appropriate range. We experimented with different ranges (Table S2 and S3 in Supporting Document S1) and chose the range with the highest accuracy: 5–15 for KEGG and 5–25 for STRING. Here we should mention that the sizes of some MISs may be below or above the range (Table 3). We did not discard the smaller MISs in the MIS generation step, as their number of probes may be sufficient. Also, the bigger MISs could not be further divided because of their special topologies, e.g., a star network or a highly dense network (Figure S2 in Supporting Document S1). We used bioDBnet [29] to map the genes or proteins to the microarray probes.\\n\\n10.1371/journal.pone.0084227.t003\\nTable 3\\n\\nMolecular Interaction Subnetwork Size.\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\nNetwork\\nAppropriate Range\\nMIS below the range\\nMIS within the range\\nMIS above the range\\n\\n\\n\\n\\nKEGG\\n5–15\\n1925\\n629\\n9\\n\\n\\nSTRING\\n5–25\\n170\\n847\\n23\\n\\n\\n\\n\\n\\n\\n\\nThe table shows the number of MISs with a total of nodes below, within, and above the appropriate range.\\n\\n\\n\\n\\n\\nFeatures Selected from Each Dataset\\nThe representative probesets selected from each dataset using KEGG and STRING are shown in Table S4 and S5 in Supporting Document S1, respectively. Figure 4 demonstrates the top ranked KEGG MIS in the Leukemia dataset containing two types of samples: Acute Myeloid Leukemia (AML), and Acute Lymphoblastic Leukemia (ALL). The MIS was taken from three overlapped subnetworks from three different pathways: 1) Focal Adhesion Pathway (id: hsa04510), 2) Adherens Junction Pathway (id: hsa04520), and 3) Bacterial Invasion of Epithelial Cells Pathway (id: hsa05100). One probe (affymetrix id: X95735_at, gene symbol: ZYX) in its representative probeset is over-expressed in AML, while the others are under-expressed.\\n\\n10.1371/journal.pone.0084227.g004\\nFigure 4\\n\\nThe top ranked KEGG MIS in the Leukemia dataset.\\nA–C) Three overlapped subnetworks from three different pathways constitute the MIS. D) The merged MIS is shown here. E) The expression heatmap of the representative probeset of the MIS is shown here.\\n\\n\\n\\n\\n\\nClassification Accuracy\\nThe LOOCV accuracies on the seven cancer datasets are shown in Table 4. COSSY achieved LOOCV accuracies of 93.2% and 92.7% using KEGG and STRING, respectively, which is comparable to the state-of-the-art classifiers, k-TSP and SVM (Support Vector Machine), and better than other classifiers (DIRAC, TSP, Doublet, PAM, Decision Tree, Nearest Neighbor, Nave Bayes) including a recent network-based approach, DIRAC [7]. Figure 5 shows the classification accuracies of five classification methods: COSSY using KEGG, COSSY using STRING, k-TSP, SVM, and DIRAC. From this figure, it is evident that COSSY’s performance on the five datasets (Leukemia, DLBCL, Prostate1, Prostate3, and Lung) is almost equal to that of k-TSP and SVM; k-TSP outperforms others using CNS, and SVM outperforms others using GCM significantly. Notably, COSSY’s performance on six datasets is significantly better than that of DIRAC; COSSY’s performance on the remaining dataset (Prostate3) is equal to that of DIRAC. On average, COSSY’s accuracy is more than 10% higher than that of DIRAC (COSSY using KEGG - 93.2%, COSSY using STRING - 92.7%, DIRAC - 82.5%). We also report the Area Under Curve (AUC) values of Receiver Operating Characteristic (ROC) curves achieved in the cross validation by COSSY and DIRAC in Table 5. COSSY outperforms DIRAC in terms of AUC (COSSY using KEGG - 0.948, COSSY using STRING - 0.945, DIRAC - 0.813). Even though a random classifier, which predicts the majority class, may have high accuracies due to the imbalance in the number of positive and negative samples in the datasets, the random classifier is unlikely to achieve as high AUCs as COSSY achieved.\\n\\n10.1371/journal.pone.0084227.g005\\nFigure 5\\n\\nLOOCV accuracy of five notable classifiers.\\nCOSSY [KEGG] and COSSY [STRING] stand for COSSY using KEGG and STRING, respectively. k-TSP and DIRAC are the classification algorithms described in [5] and [17], respectively. SVM stands for the Support Vector Machine algorithm.\\n\\n\\n\\n\\n10.1371/journal.pone.0084227.t004\\nTable 4\\n\\nLOOCV accuracy (%) of classifiers.\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\nMethod\\nLeukemia\\nCNS\\nDLBCL\\nProstate1\\nProstate3\\nLung\\nGCM\\nAverage\\n\\n\\n\\n\\nCOSSY [KEGG]\\n98.6\\n85.3\\n93.5\\n90.2\\n100.0\\n99.5\\n85.0\\n93.2\\n\\n\\nCOSSY [STRING]\\n95.8\\n88.2\\n94.8\\n90.2\\n97.0\\n98.3\\n84.6\\n92.7\\n\\n\\nDIRAC\\n94.8\\n72.3\\n73.4\\n62.9\\n100.0\\n98.8\\n75.2\\n82.5\\n\\n\\n\\nk-TSP*\\n\\n95.8\\n97.1\\n97.4\\n91.2\\n97.0\\n98.9\\n85.4\\n94.7\\n\\n\\nTSP*\\n\\n93.8\\n77.9\\n98.1\\n95.1\\n97.0\\n98.3\\n75.4\\n90.8\\n\\n\\nSVM*\\n\\n98.6\\n82.4\\n97.4\\n91.2\\n100.0\\n99.5\\n93.2\\n94.6\\n\\n\\nDoublet [Sign-DT]+\\n\\n93.1\\n82.4\\n97.4\\n86.3\\n97.0\\n98.3\\n85.0\\n91.3\\n\\n\\nDoublet [Sumdiff-DT]+\\n\\n91.7\\n70.6\\n97.4\\n82.4\\n87.9\\n95.0\\n81.4\\n86.6\\n\\n\\nDoublet [Mul-DT]+\\n\\n84.7\\n55.9\\n97.4\\n86.3\\n90.9\\n92.3\\n83.2\\n84.4\\n\\n\\nDecision Tree (DT)*\\n\\n73.6\\n67.7\\n80.5\\n87.3\\n84.9\\n96.1\\n77.9\\n81.1\\n\\n\\nNave Bayes*\\n\\n100.0\\n82.4\\n80.5\\n62.8\\n90.9\\n97.8\\n84.3\\n85.5\\n\\n\\n\\nk Nearest Neighbor*\\n\\n84.7\\n76.5\\n84.4\\n76.5\\n87.9\\n98.3\\n82.9\\n84.5\\n\\n\\nPAM*\\n\\n97.2\\n82.4\\n85.7\\n91.2\\n100.0\\n99.5\\n79.3\\n90.7\\n\\n\\n\\n\\n\\n\\n\\nThe leftmost column contains the names of the methods; the rightmost column shows the average accuracy of each method for seven datasets, and other columns show the accuracy (%) for individual datasets. ‘COSSY [KEGG]’ and ‘COSSY [STRING]’ represent COSSY using KEGG and STRING, respectively. ‘DIRAC’ is the algorithm proposed in [17] whose LOOCV accuracies have been calculated using the matlab code published with the paper. k-TSP and TSP denote the classification algorithms described in [5] and [4], respectively. SVM stands for Support Vector Machine. ‘Doublet [Sign-DT]’, ‘Doublet [Sumdiff-DT]’, and ‘Doublet [Mul-DT]’ denote the classification methods using Sign-Doublet, Sumdiff-Doublet, and Mul-Doublet, respectively, with decision trees as described in [6]. The last three rows contain the loocv accuracies using Nave Bayes, k Nearest Neighbor, and PAM classifier, respectively.\\n\\n\\n\\nResults obtained from [5].\\n\\n\\n\\nResults obtained from [6].\\n\\n\\n\\n\\n10.1371/journal.pone.0084227.t005\\nTable 5\\n\\nArea Under Curve (AUC) values of COSSY and DIRAC for different datasets.\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\nDataset Name\\nAUC of COSSY using KEGG\\nAUC of COSSY using STRING\\nAUC of DIRAC\\n\\n\\n\\n\\nLeukemia\\n\\n0.986\\n\\n0.985\\n0.948\\n\\n\\nCNS\\n0.862\\n\\n0.876\\n\\n0.726\\n\\n\\nDLBCL\\n\\n0.985\\n\\n0.976\\n0.636\\n\\n\\nProstate1\\n0.909\\n\\n0.918\\n\\n0.635\\n\\n\\nProstate3\\n\\n1.000\\n\\n0.972\\n\\n1.000\\n\\n\\n\\nLung\\n\\n0.999\\n\\n\\n0.999\\n\\n0.990\\n\\n\\nGCM\\n\\n0.896\\n\\n0.889\\n0.757\\n\\n\\nAverage\\n\\n0.948\\n\\n0.945\\n0.813\\n\\n\\n\\n\\n\\n\\n\\nAUC has been calculated using the ROCR package in R [35]. The best AUC for each dataset is highlighted in bold face.\\n\\n\\n\\n\\n\\nInterpretation of the COSSY Results on the Leukemia Dataset\\nAs we discussed in this paper, the advantage of COSSY is its ability to automatically discover the important subnetworks to differentiate between two phenotypes (context-specific). We used the Leukemia dataset to illustrate this interpretation of the subnetworks identified by COSSY. As demonstrated in the results section, we selected the top five differentially expressed genes from each subnetwork for classification. Not surprisingly, the top subnetwork identified by COSSY contains ZYX (zyxin), which has been previously identified as the top differentially expressed gene to distinguish AML from ALL in this dataset [20]. Other genes previously identified as gene signatures to distinguish AML from ALL include CTSD, LYN, MYB [20].\\nExamining the over-expressed genes in AML across the top 15 subnetworks (Table S4 in Supporting Document S1), COSSY identified LYN/PI3K/AKT signaling (LYN, PIK3R2, AKT1), lysosome complex (LAMP2, CTSD, ATP6AP1), and integrin signaling pathways (ITGAX, ITGB2, FCER1G) as key pathways in driving this disease. By querying the core genes identified from the top MISs to the KEGG pathways, we could ‘stitch’ these subnetworks (MISs) together as illustrated in Figure 6. This ‘stitched’ network revealed three key pathways that have been described to be deregulated in AML. In particular, LYN, an Src-family kinase member, has been found to be over-expressed in AML [20], [30]. Previous studies described LYN to be linked to mTOR [30]; we revealed this link between LYN and mTOR via PIK3R2 and AKT1 by stitching these subnetworks together in this analysis.\\n\\n10.1371/journal.pone.0084227.g006\\nFigure 6\\n\\nStitching of the MISs found from the Leukemia dataset.\\nThe number at the end of the name of an MIS indicates its rank.\\n\\n\\n\\nPI3K/AKT signaling has been implicated as playing a critical role in AML [30]–[32]; also, inhibiting multiple components of this pathway might provide the therapeutic interventions to this lethal disease. From the stitched network, we identified ITGAX and ITGB2 as over-expressed in AML; these two integrin receptors interact with each other as well as with FCER1G. Recently, it has been found that the interactions of ITGAX/ITGB2 and FCER1G could signal SYK-dependent activation of the JAK/STAT pathway in AML, and this signaling axis might serve as a novel therapeutic target for AML [33].\\nFinally, we observed that the multiple components of the lysosome (CTSD, LAMP2, and ATP6AP1) were over-expressed in AML. This is an interesting finding as a recent paper demonstrated that using mefloquine, an anti-malaria drug, could disrupt lysosome in AML and increase cancer cell killing effects [34]. Furthermore, AML cells were less viable due to gene knock-down on LAMP2. This demonstrates that lysosome could be a new therapeutic target in AML [34]. From this example, we highlight that COSSY can identify interesting and context-specific subnetworks from microarray gene expression datasets. By stitching the core genes together from these subnetworks, novel biology could be discovered. This represents an added value to COSSY as compared to static pathway-based analyses such as Gene Set Enrichment Analysis.\\n\\n\\n\\nDiscussion\\n\\nGene Interaction\\nConventional differential expression analyses focus on single gene markers. Even the gene pairs identified by k-TSP or doublets do not necessarily imply a biochemical interaction between the genes. COSSY, in contrast, takes biological interactions under consideration ensuring a differential expression pattern of the representative probeset instead of a differential expression of a single gene. Consequently, it provides an opportunity to discover the development process of a phenotype.\\n\\n\\nNetwork Partition\\nUnlike GSEA and DIRAC, COSSY does not require any pre-partitioned subnetworks and is not limited by the quality of the predefined gene sets. COSSY can automatically generate the subnetworks. Importantly, these subnetworks may also be used with other types of analyses requiring pre-partitioned subnetworks, which is not yet possible from any of the existing subnetwork discovery methods.\\n\\n\\nNon-greedy Approach\\nThe existing subnetwork discovery methods [13]–[16] usually build subnetworks using a greedy approach that starts with a small network (typically a single node) and then adds more nodes greedily until the addition improves the quality. However, it may have local-optima problems. On the other hand, as our method follows the non-greedy approach – divides the network first and then calculates its effectiveness – local-optima problems are unlikely to occur. Moreover, the applicability of the previous greedy methods often depends on the network topology, which is explained in the following section.\\n\\n\\nCOSSY is Independent of Network Topology\\nCurrent subnetwork discovery methods are highly dependent on the network topology. The method by Chuang et al. (2007) [13] is likely to stop at a locally optimum node frequently where strict linear paths dominate in a network such as KEGG. Chowdhury et al. (2010) [14] use the set-cover-based algorithm where either all the positive or negative samples have to be covered. If the network contains small components, then it might not be able to cover the positive or negative samples completely. Dao et al. (2010) [15] utilize density-constrained clustering which would not work well on sparse networks such as KEGG. Su et al. (2010) [16] look for the discriminative linear paths and then combine them to build the subnetwork. In a sparse network, the linear paths may not intersect each other. In contrast, we showed that COSSY works well on networks with highly different topologies such as KEGG and STRING. The average LOOCV accuracies of COSSY are 93.2% and 92.7% when used with KEGG and STRING, respectively, and are high and close to each other.\\n\\n\\nSubnetworks with Heterogeneous Characteristics\\nChuang et al. (2007) [13] take the sum of the expressions of all the nodes in a subnetwork to aggregate its expressions, and thus the process favors the subnetworks with homogeneous characteristics where all the nodes are either positively or negatively expressed. The NetCover algorithm [14] also aggregates the expressions in the same way and suffers from the same problem. Instead of aggregating the expressions, COSSY uses clustering and thus it is capable of utilizing the heterogeneous nature to find the subnetworks that have both positively and negatively expressed nodes. For example, the representative probeset of the Leukemia MIS in Figure 4 contains one positively and four negatively expressed probes.\\n\\n\\n\\nConclusion\\nIn this paper, we have introduced a non-greedy algorithm, COSSY, to partition a molecular interaction network and find the important subnetworks discriminating between two phenotypes. COSSY can find subnetworks with heterogeneous characteristics in any network irrespective of the topology. Its accuracy is comparable to the state-of-the-art classifiers. We also illustrated an interpretation of the results to discover the development process of a disease.\\n\\n\\nSupporting Information\\n\\nDocument S1\\n\\n\\nSupporting information document. This document contains seven sections (S1. Estimated t-score, S2. Binary Vote, S3. Dataset Download Sources, S4. An Experiment with the appropriate range, S5. Algorithm Illustration, S6. Irregular Network Topology, and S7. Representative probesets of the top MISs), two figures (S1. An illustration of the MIS generation, and S2. Topology of out-of-size MISs), and five Tables (S1. Dataset Download Sources, S2. LOOCV accuracy with different appropriate ranges from KEGG, S3. LOOCV accuracy with different appropriate ranges from STRING, S4. Representative probesets of the top MISs from KEGG, and S5. Representative probesets of the top MISs from STRING).\\n(PDF)\\n\\n\\n\\nClick here for additional data file.\\n\\n\\n\\n\\n\\n\\nReferences\\n\\n1\\n\\nRischNJ (2000) Searching for genetic determinants in the new millennium. Nature\\n405: 847–856.10866211\\n\\n\\n2\\n\\nMarchiniJ, DonnellyP, CardonLR (2005) Genome-wide strategies for detecting multiple loci that influence complex diseases. Nature Genetics\\n37: 413–417.15793588\\n\\n\\n3\\n\\nSchadtEE (2009) Molecular networks as sensors and drivers of common human diseases. Nature\\n461: 218–223.19741703\\n\\n\\n4\\n\\nGemanD, D’AvignonC, NaimanDQ, WinslowRL (2004) Classifying gene expression profiles from pairwise mRNA comparisons. Statistical Applications in Genetics and Molecular Biology\\n3: Article19.16646797\\n\\n\\n5\\n\\nTanAC, NaimanDQ, XuL, WinslowRL, GemanD (2005) Simple decision rules for classifying human cancers from gene expression profiles. Bioinformatics\\n21: 3896–3904.16105897\\n\\n\\n6\\n\\nChopraP, LeeJ, KangJ, LeeS (2010) Improving cancer classification accuracy using gene pairs. PLoS ONE\\n5: e14305.21200431\\n\\n\\n7\\n\\nLinX, AfsariB, MarchionniL, CopeL, ParmigianiG, et al (2009) The ordering of expression among a few genes can provide simple cancer biomarkers and signal BRCA1 mutations. BMC bioinformatics\\n10: 256.19695104\\n\\n\\n8\\n\\nMagisAT, PriceND (2012) The top-scoring ‘N’ algorithm: a generalized relative expression classification method from small numbers of biomolecules. BMC bioinformatics\\n13: 227.22966958\\n\\n\\n9\\n\\nWangH, ZhangH, DaiZ, ChenMs, YuanZ (2013) TSG: a new algorithm for binary and multi-class cancer classification and informative genes selection. BMC medical genomics\\n6 Suppl 1S3.\\n\\n\\n10\\n\\nSubramanianA, TamayoP, MoothaVK, MukherjeeS, EbertBL, et al (2005) Gene set enrichment analysis: A knowledge-based approach for interpreting genome-wide expression profiles. Proceedings of the National Academy of Sciences of the United States of America\\n102: 15545–15550.16199517\\n\\n\\n11\\n\\nKanehisaM, GotoS, SatoY, FurumichiM, TanabeM (2012) KEGG for integration and interpretation of large-scale molecular data sets. Nucleic acids research\\n40: D109–14.22080510\\n\\n\\n12\\n\\nFranceschiniA, SzklarczykD, FrankildS, KuhnM, SimonovicM, et al (2013) STRING v9.1: protein-protein interaction networks, with increased coverage and integration. Nucleic acids research\\n41: D808–15.23203871\\n\\n\\n13\\n\\nChuangHY, LeeE, LiuYT, LeeD, IdekerT (2007) Network-based classification of breast cancer metastasis. Molecular Systems Biology\\n3: 140.17940530\\n\\n\\n14\\n\\nChowdhurySA, KoyutürkM (2010) Identification of coordinately dysregulated subnetworks in complex phenotypes. Pacific Symposium On Biocomputing\\n144: 133–144.\\n\\n\\n15\\n\\nDaoP, ColakR, SalariR, MoserF, DavicioniE, et al (2010) Inferring cancer subnetwork markers using density-constrained biclustering. Bioinformatics\\n26: i625–i631.20823331\\n\\n\\n16\\n\\nSuJ, YoonBJ, DoughertyER (2010) Identification of diagnostic subnetwork markers for cancer in human protein-protein interaction network. BMC Bioinformatics\\n11: S8.\\n\\n\\n17\\n\\nEddyJA, HoodL, PriceND, GemanD (2010) Identifying Tightly Regulated and Variably Expressed Networks by Differential Rank Conservation (DIRAC). PLoS Computational Biology\\n6: 17.\\n\\n\\n18\\n\\nClausetA, NewmanMEJ, MooreC (2004) Finding community structure in very large networks. Physical Review E\\n70: 1–6.\\n\\n\\n19\\n\\nStalteriMA, HarrisonAP (2007) Interpretation of multiple probe sets mapping to the same gene in Affymetrix GeneChips. BMC Bioinformatics\\n8: 13.17224057\\n\\n\\n20\\n\\nGolubTR, SlonimDK, TamayoP, HuardC, GaasenbeekM, et al (1999) Molecular classification of cancer: class discovery and class prediction by gene expression monitoring. Science\\n286: 531–537.10521349\\n\\n\\n21\\n\\nPomeroySL, TamayoP, GaasenbeekM, SturlaLM, AngeloM, et al (2002) Prediction of central nervous system embryonal tumour outcome based on gene expression. Nature\\n415: 436–442.11807556\\n\\n\\n22\\nShipp MA, Ross KN, Tamayo P, Weng AP, Kutok JL, et al.. (2002) Diffuse large B-cell lymphoma outcome prediction by gene-expression profiling and supervised machine learning. Technical Report 1, Dana-Farber Cancer Institute, Harvard Medical School, Boston, Massachusetts, USA. margaret shipp@dfci.harvard.edu.\\n\\n\\n23\\n\\nSinghD, FebboPG, RossK, JacksonDG, ManolaJ, et al (2002) Gene expression correlates of clinical prostate cancer behavior. Cancer Cell\\n1: 203–209.12086878\\n\\n\\n24\\n\\nWelshJB, SapinosoLM, SuAI, KernSG, Wang-rodriguezJ, et al (2001) Analysis of Gene Expression Identifies Candidate Markers and Pharmacological Targets in Prostate Cancer. Cancer Research\\n61: 5974–5978.11507037\\n\\n\\n25\\n\\nGordonGJ, JensenRV, HsiaoLL, GullansSR, BlumenstockJE, et al (2002) Translation of microarray data into clinically relevant cancer diagnostic tests using gene expression ratios in lung cancer and mesothelioma. Cancer Research\\n62: 4963–4967.12208747\\n\\n\\n26\\n\\nRamaswamyS, TamayoP, RifkinR, MukherjeeS, YeangCH, et al (2001) Multiclass cancer diagnosis using tumor gene expression signatures. Proceedings of the National Academy of Sciences of the United States of America\\n98: 15149–15154.11742071\\n\\n\\n27\\n\\nWardJH (1963) Hierarchical grouping to optimize an objective function. Journal of the American Statistical Association\\n58: 236–244.\\n\\n\\n28\\nMardia KV, Kent JT, Bibby JM (1979) Multivariate analysis. Academic Press.\\n\\n\\n29\\n\\nMudunuriU, CheA, YiM, StephensRM (2009) bioDBnet: the biological database network. Bioinformatics\\n25: 555–556.19129209\\n\\n\\n30\\n\\nDos SantosC, DemurC, BardetV, Prade-HoudellierN, PayrastreB, et al (2008) A critical role for Lyn in acute myeloid leukemia. Blood\\n111: 2269–79.18056483\\n\\n\\n31\\n\\nParkS, ChapuisN, TamburiniJ, BardetV, Cornillet-LefebvreP, et al (2010) Role of the PI3K/AKT and mTOR signaling pathways in acute myeloid leukemia. Haematologica\\n95: 819–28.19951971\\n\\n\\n32\\n\\nTamburiniJ, ChapuisN, BardetV, ParkS, SujobertP, et al (2008) Mammalian target of rapamycin (mTOR) inhibition activates phosphatidylinositol 3-kinase/Akt by up-regulating insulinlike growth factor-1 receptor signaling in acute myeloid leukemia: rationale for therapeutic inhibition of both pathways. Blood\\n111: 379–82.17878402\\n\\n\\n33\\n\\nOellerichT, OellerichMF, EngelkeM, MünchS, MohrS, et al (2013) \\nβ2 integrin-derived signals induce cell survival and proliferation of AML blasts by activating a Syk/STAT signaling axis. Blood\\n121: 3889–99.23509157\\n\\n\\n34\\n\\nSukhaiMA, PrabhaS, HurrenR, RutledgeAC, LeeAY, et al (2013) Lysosomal disruption preferentially targets acute myeloid leukemia cells and progenitors. The Journal of clinical investigation\\n123: 315–28.23202731\\n\\n\\n35\\n\\nSingT, SanderO, BeerenwinkelN, LengauerT (2005) ROCR: visualizing classifier performance in R. Bioinformatics (Oxford, England). 21: 3940–1.\\n\\n\\n\\n\\n\""
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "page_soup.get_text()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<date date-type=\"received\">\n",
       " <day>10</day>\n",
       " <month>9</month>\n",
       " <year>2013</year>\n",
       " </date>, <date date-type=\"accepted\">\n",
       " <day>21</day>\n",
       " <month>11</month>\n",
       " <year>2013</year>\n",
       " </date>]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "page_soup.findAll('date')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Affiliations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<aff id=\"aff1\">\n",
       " <label>1</label>\n",
       " <addr-line>Department of Computer Science and Engineering, Korea University, Seoul, Korea</addr-line>\n",
       " </aff>, <aff id=\"aff2\">\n",
       " <label>2</label>\n",
       " <addr-line>Department of Medicine/Medical Oncology, University of Colorado Anschutz Medical Campus, Aurora, Colorado, United States of America</addr-line>\n",
       " </aff>, <aff id=\"aff3\">\n",
       " <label>3</label>\n",
       " <addr-line>Interdisciplinary Graduate Program in Bioinformatics, Korea University, Seoul, Korea</addr-line>\n",
       " </aff>, <aff id=\"edit1\">\n",
       " <addr-line>Institute for Research in Biomedicine, Spain</addr-line>\n",
       " </aff>]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "page_soup.findAll('aff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<aff id=\"aff2\">\n",
       "<label>2</label>\n",
       "<addr-line>Department of Medicine/Medical Oncology, University of Colorado Anschutz Medical Campus, Aurora, Colorado, United States of America</addr-line>\n",
       "</aff>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "page_soup.find('aff', {'id':'aff2'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. See topics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<subject>Research Article</subject>\n"
     ]
    }
   ],
   "source": [
    "print(page_soup.find('subject'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<subject>Research Article</subject>,\n",
       " <subject>Biology</subject>,\n",
       " <subject>Computational Biology</subject>,\n",
       " <subject>Molecular Genetics</subject>,\n",
       " <subject>Gene Classes</subject>,\n",
       " <subject>Gene Identification and Analysis</subject>,\n",
       " <subject>Gene Expression</subject>,\n",
       " <subject>Metabolic Networks</subject>,\n",
       " <subject>Microarrays</subject>,\n",
       " <subject>Regulatory Networks</subject>,\n",
       " <subject>Signaling Networks</subject>,\n",
       " <subject>Genetics</subject>,\n",
       " <subject>Human Genetics</subject>,\n",
       " <subject>Genetic Association Studies</subject>,\n",
       " <subject>Gene Networks</subject>,\n",
       " <subject>Genetics of Disease</subject>,\n",
       " <subject>Computer Science</subject>,\n",
       " <subject>Algorithms</subject>]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "page_soup.findAll('subject')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. All tags"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['front', 'sec', 'sec', 'sec', 'sec', 'sec', 'sec', 'back']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "soup_temp = page_soup.html.body.findChild('pmc-articleset').article\n",
    "[tag.name for tag in soup_temp.find_all(recursive=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{},\n",
       " {'id': 's1'},\n",
       " {'id': 's2'},\n",
       " {'id': 's3'},\n",
       " {'id': 's4'},\n",
       " {'id': 's5'},\n",
       " {'sec-type': 'supplementary-material', 'id': 's6'},\n",
       " {}]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[tag.attrs for tag in soup_temp.find_all(recursive=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'xmlns:xlink': 'http://www.w3.org/1999/xlink',\n",
       " 'xmlns:mml': 'http://www.w3.org/1998/Math/MathML',\n",
       " 'article-type': 'research-article'}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "soup_temp.attrs"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
